<!-- Generated by staticdocs: do not edit by hand -->
<html>
  <head>
  <meta charset="utf-8">
<title>Home. valr</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script   src="https://code.jquery.com/jquery-3.1.0.min.js"   integrity="sha256-cCueBR6CsyA4/9szpPfrX3s49M9vUU5BgtiJj06wt/s="   crossorigin="anonymous"></script>

<!-- Latest compiled and minified CSS -->
<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">


<!-- Latest compiled and minified JavaScript -->
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- from staticdocs -->
<link href="css/highlight.css" rel="stylesheet">
<link href="css/staticdocs.css" rel="stylesheet">

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$','$'], ["\\(","\\)"] ],
      processEscapes: true
    }
  });
</script>
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container">
        <header>
        
<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="articles/valr.html">Overview</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
        
        </header>

        <div class="row">
  <div class="col-md-9">
    <h1>valr</h1>

<p><a href="https://travis-ci.org/jayhesselberth/valr"><img src="https://travis-ci.org/jayhesselberth/valr.svg?branch=master" alt="Build Status"/></a>
<a href="https://codecov.io/github/jayhesselberth/valr?branch=master"><img src="https://img.shields.io/codecov/c/github/jayhesselberth/valr/master.svg" alt="Coverage Status"/></a></p>

<p><code>valr</code> provides tools for fast, exploratory analysis of large biological data sets <strong>within R/RStudio</strong>. <code>valr</code> was developed from a desire to teach the powerful concepts of genome interval arithmetic, but without the cumbersome back-and-forth between command-line and exploratory analysis tools. </p>

<p>Key parts of <code>valr</code> are implemented in <a href="http://www.rcpp.org/"><code>Rcpp</code></a> for speed. Moreover, <code>valr</code> integrates with <a href="https://github.com/hadley/dplyr"><code>dplyr</code></a> and the <code>magrittr</code> pipe operator (<code>%&gt;%</code>) for an expressive syntax and can be used in reproducible reports written in <a href="http://rmarkdown.rstudio.com/"><code>RMarkdown</code></a>.</p>

<p>See the <a href="http://jayhesselberth.github.io/valr"><code>valr</code> documentation</a> for a complete API reference and examples.</p>

<h2>Installation</h2>

<p><code>valr</code> can be installed from github:</p>

<pre><code class="R">devtools::install_github(&#39;jayhesselberth/valr&#39;)
</code></pre>

<p><strong>Note</strong> that <code>valr</code> requires a full-featured C++11 compiler (<code>gcc&gt;=6.0</code> or <code>clang++</code>)</p>

<h2>API</h2>

<p>Function names are similar to their their <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> counterparts, with some additions.</p>

<h3>Reading data</h3>

<ul>
<li><p>BED and related files are read with <code>read_bed()</code>, <code>read_bed12()</code>, <code>read_bedgraph()</code>, <code>read_narrowpeak()</code> and <code>read_broadpeak()</code>.</p></li>
<li><p>Genome files containing chromosome name and size information are loaded with <code>read_genome()</code>.</p></li>
<li><p>VCF files are loaded with <code>read_vcf()</code>.</p></li>
</ul>

<h3>Transforming single interval sets</h3>

<ul>
<li><p>Intervals are ordered with <code>bed_sort()</code>.</p></li>
<li><p>Interval coordinates are adjusted with <code>bed_slop()</code> and <code>bed_shift()</code>, and new flanking intervals are created with <code>bed_flank()</code>.</p></li>
<li><p>Nearby intervals are combined with <code>bed_merge()</code> and identified (but not merged) with <code>bed_cluster()</code>.  </p></li>
<li><p>Intervals not covered by a query are created with <code>bed_complement()</code>.</p></li>
</ul>

<h3>Comparing multiple interval sets</h3>

<ul>
<li><p>Find overlaps between two sets of intervals with <code>bed_intersect()</code>.</p></li>
<li><p>Apply functions to selected columns for overlapping intervals with <code>bed_map()</code>.</p></li>
<li><p>Remove intervals based on overlaps between two files with <code>bed_subtract()</code>.</p></li>
<li><p>Find overlapping intervals within a window with <code>bed_window()</code>.</p></li>
<li><p>Find the closest intervals independent of overlaps with <code>bed_closest()</code>.</p></li>
</ul>

<h3>Randomizing intervals</h3>

<ul>
<li><p>Generate random intervals from an input genome with <code>bed_random()</code>.</p></li>
<li><p>Shuffle the coordinates of input intervals with <code>bed_shuffle()</code>.</p></li>
<li><p>Random sampling of input intervals is done with the <code>sample_</code> function family in <code>dplyr</code>.</p></li>
</ul>

<h3>Interval statistics</h3>

<ul>
<li><p>Quality overlaps between two sets of intervals with <code>bed_fisher()</code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code>bed_reldist()</code> and <code>bed_absdist()</code>.</p></li>
<li><p>Quantify extent of overlap between two sets of intervals with <code>bed_jaccard()</code>.</p></li>
</ul>

<h2>Related work</h2>

<ul>
<li><p>Command-line tools <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> and <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>.</p></li>
<li><p>The Python library <a href="https://pythonhosted.org/pybedtools/">pybedtools</a> wraps BEDtools.</p></li>
<li><p>The R packages <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://cran.r-project.org/web/packages/bedr/index.html">bedr</a>, <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a> and <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">GenometriCorr</a> provide similar capability with a different philosophy.</p></li>
</ul>

  </div>
</div>

        <footer>
        <p>Built by <a href="http://hadley.github.io/staticdocs/">staticdocs</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
        </footer>
   </div>

  </body>
</html>
