<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title><code>valr</code>: Genome interval arithmetic in R. valr</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container">
      <header><div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/valr-overview.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1><code>valr</code>: Genome interval arithmetic in R</h1>
                        <h4 class="author">Jay Hesselberth <a href="mailto:jay.hesselberth@gmail.com">jay.hesselberth@gmail.com</a></h4>
            
            <h4 class="date">2016-10-12</h4>
          </div>

    
    
<div id="overview" class="section level1">
<h1>Overview</h1>
<p><code>valr</code> provides tools to read and manipulate genome intervals and signals, similar to the standalone <a href="http://bedtools.readthedocs.org/en/latest/"><code>BEDtools</code></a> suite. <code>valr</code> enables BEDtools-like analysis in the R/RStudio environment, and uses modern R tools for a terse, expressive syntax. Compute-intensive algorithms are implemented in <a href="http://www.rcpp.org/"><code>Rcpp</code></a>/C++, and many methods take advantage of the speed and grouping capability provided by <a href="https://github.com/hadley/dplyr"><code>dplyr</code></a>.</p>
<p>We provide several introductions to <code>valr</code>:</p>
<ul><li>This comprehensive <code>vignette</code> covering the core methods.</li>
<li>A tutorial that demonstrates how to use <code>valr</code> in &ldquo;real-world&rdquo; applications.</li>
<li>A <code>shiny</code> application that demonstrates interactive analysis of genome-scale data sets.</li>
</ul><div id="installation" class="section level2">
<h2>Installation</h2>
<p><code>valr</code> can be installed from github, and will be eventually deposited in CRAN.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">'jayhesselberth/valr'</span>)</code></pre></div>
</div>
<div id="comparison-to-other-tools" class="section level2">
<h2>Comparison to other tools</h2>
<p><strong>Why another tool set for interval manipulations?</strong> We already have <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a>, <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>, <a href="https://pythonhosted.org/pybedtools/">pybedtools</a>, <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://cran.r-project.org/web/packages/bedr/index.html">bedr</a> and <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a>.</p>
<p>We were motivated to develop a toolset that:</p>
<ul><li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the &ldquo;modern R&rdquo; syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code>%&gt;%</code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visulaizations with <a href="http://shiny.rstudio.com/"><code>shiny</code></a>.</li>
</ul><p>We anticipate <code>valr</code> will mainly be used for analysis of pre-processed data in BED, bedGraph and VCF formats. Most users will have processed their aligned reads from BAM format to bedGraph, so we do not foresee supporting BAM directly. We would entertain requests for GTF / GFF support if there is interest, as <code>tidyr</code> makes it easy to convert these to BED12.</p>
<p>Certain algorithms in <code>valr</code> were implemented in <code>Rcpp</code> (including <a href="https://github.com/jayhesselberth/valr/tree/master/src">intersect, merge, subtract, closest</a>) to enable fluid interactive analysis. See the <a href="#benchmarks">benchmarks</a> section for details.</p>
</div>
<div id="non-standard-evaluation" class="section level2">
<h2>Non-standard evaluation</h2>
<p>Several of the methods in <code>valr</code> use <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html">NSE</a> for an expressive syntax. Columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">mean =</span> <span class="kw">mean</span>(value), <span class="dt">var =</span> <span class="kw">var</span>(value))
<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x, <span class="dt">concat =</span> <span class="kw"><a href="../reference/bed_map.html">concat</a></span>(value), <span class="dt">max =</span> <span class="kw">max</span>(value))</code></pre></div>
</div>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Here is an example using <code>valr</code> that creates</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)</code></pre></div>
</div>
</div>
<div id="reading-data" class="section level1 tabset">
<h1>Reading data</h1>
<p><code>valr</code> has several methods to read interval data. These methods:</p>
<ul><li>Take local files or URLs as input.</li>
<li>Return a <code>data.frame</code> in <code>tibble::tbl_df</code> format.</li>
<li>Assign consistent <code>chrom</code>, <code>start</code> and <code>end</code> column names.</li>
<li>Use <a href="https://github.com/hadley/readr"><code>readr</code></a> for speed.</li>
<li>Coerce column types.</li>
</ul><p>The methods include:</p>
<ul><li><code><a href="../reference/read_bed.html">read_bed()</a></code>: read a BED3+ file</li>
<li><code><a href="../reference/read_bed.html">read_bed12()</a></code>: read a BED12 file</li>
<li><code><a href="../reference/read_bed.html">read_bedgraph()</a></code>: read a bedGraph file</li>
<li><code><a href="../reference/read_genome.html">read_genome()</a></code>: read a UCSC &ldquo;chrom size&rdquo; file</li>
<li><code><a href="../reference/read_vcf.html">read_vcf()</a></code>: read the Variant Call Format</li>
<li><code><a href="../reference/read_bed.html">read_narrowpeak()</a></code>: read narrowPeak files</li>
<li><code><a href="../reference/read_bed.html">read_broadpeak()</a></code>: read broadPeak files</li>
</ul><div id="bed-files" class="section level2">
<h2>BED files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'3fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 3</span>
<span class="co">#&gt;    chrom  start    end</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409</span>
<span class="co">#&gt; 2   chr1  14361  19759</span>
<span class="co">#&gt; 3   chr1  14406  29370</span>
<span class="co">#&gt; 4   chr1  34610  36081</span>
<span class="co">#&gt; 5   chr1  69090  70008</span>
<span class="co">#&gt; 6   chr1 134772 140566</span>
<span class="co">#&gt; 7   chr1 321083 321115</span>
<span class="co">#&gt; 8   chr1 321145 321207</span>
<span class="co">#&gt; 9   chr1 322036 326938</span>
<span class="co">#&gt; 10  chr1 327545 328439</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="dt">n_fields =</span> <span class="dv">6</span>, <span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'6fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409      DDX11L1     3      +</span>
<span class="co">#&gt; 2   chr1  14361  19759       WASH7P    10      -</span>
<span class="co">#&gt; 3   chr1  14406  29370       WASH7P     7      -</span>
<span class="co">#&gt; 4   chr1  34610  36081      FAM138F     3      -</span>
<span class="co">#&gt; 5   chr1  69090  70008        OR4F5     1      +</span>
<span class="co">#&gt; 6   chr1 134772 140566    LOC729737     3      -</span>
<span class="co">#&gt; 7   chr1 321083 321115     DQ597235     1      +</span>
<span class="co">#&gt; 8   chr1 321145 321207     DQ599768     1      +</span>
<span class="co">#&gt; 9   chr1 322036 326938 LOC100133331     3      +</span>
<span class="co">#&gt; 10  chr1 327545 328439    LOC388312     1      +</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed12</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'12fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 3 &times; 12</span>
<span class="co">#&gt;   chrom   start     end      name score strand cds_start cds_end item_rgb</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;   &lt;int&gt;    &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1 4797973 4836816  testgene     1      +   4797973 4836816        .</span>
<span class="co">#&gt; 2 chr10 4848118 4880877 diffchrom     1      +   4848118 4880877        .</span>
<span class="co">#&gt; 3 chr20 5073253 5152630 negstrand     1      -   5073253 5152630        .</span>
<span class="co">#&gt; # ... with 3 more variables: exon_count &lt;int&gt;, exon_sizes &lt;chr&gt;,</span>
<span class="co">#&gt; #   exon_starts &lt;chr&gt;</span></code></pre></div>
</div>
<div id="genome-files" class="section level2">
<h2>Genome files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome_human &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>))</code></pre></div>
</div>
<div id="bedgraph-files" class="section level2">
<h2>bedGraph files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.bg.gz'</span>))
<span class="co">#&gt; # A tibble: 4 &times; 4</span>
<span class="co">#&gt;   chrom    start      end value</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr19 49302000 49302300 -1.00</span>
<span class="co">#&gt; 2 chr19 49302300 49302600 -0.75</span>
<span class="co">#&gt; 3 chr19 49302600 49302900 -0.50</span>
<span class="co">#&gt; 4 chr19 49302900 49303200 -0.25</span></code></pre></div>
</div>
<div id="vcf-files" class="section level2">
<h2>VCF files</h2>
<p><code><a href="../reference/read_vcf.html">read_vcf()</a></code> reads VCF files and assigns <code>chrom</code>, <code>start</code> and <code>end</code> columns to be used to downstream interval comparisons. Note the interval size is calculated as the length of the <code>REF</code> field in the original file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_vcf.html">read_vcf</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.vcf.gz'</span>))
<span class="co">#&gt; # A tibble: 11 &times; 18</span>
<span class="co">#&gt;    CHROM   POS      ID   REF   ALT  QUAL FILTER</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1      1    10    1:10     A     T   100   PASS</span>
<span class="co">#&gt; 2      1    20    1:20     G     C   100   PASS</span>
<span class="co">#&gt; 3      1    30    1:30     C     A   100   PASS</span>
<span class="co">#&gt; 4      1    40    1:40     A     C   100   PASS</span>
<span class="co">#&gt; 5      1 10000 1:10000     G     C   100   PASS</span>
<span class="co">#&gt; 6      1 20000 1:20000     T     A   100   PASS</span>
<span class="co">#&gt; 7      4  5000  4:5000     A     T   100   PASS</span>
<span class="co">#&gt; 8      4  6000  4:6000     C     T   100   PASS</span>
<span class="co">#&gt; 9      X   800   X:800     A     C   100   PASS</span>
<span class="co">#&gt; 10     X   900   X:900     A     T   100   PASS</span>
<span class="co">#&gt; 11     X  1000  X:1000     T     G   100   PASS</span>
<span class="co">#&gt; # ... with 11 more variables: INFO &lt;chr&gt;, FORMAT &lt;chr&gt;, X1 &lt;chr&gt;,</span>
<span class="co">#&gt; #   X2 &lt;chr&gt;, X3 &lt;chr&gt;, X4 &lt;chr&gt;, X5 &lt;chr&gt;, X6 &lt;chr&gt;, chrom &lt;chr&gt;,</span>
<span class="co">#&gt; #   start &lt;int&gt;, end &lt;int&gt;</span></code></pre></div>
</div>
</div>
<div id="interval-manipulations" class="section level1">
<h1>Interval manipulations</h1>
<p><code>valr</code> implements several methods for manipulating sets of intervals. Some methods operate on a single set of intervals, while others compare two sets of intervals.</p>
<p>Many methods the same name as the corresponding <code>BEDtool</code>, and some commonly used <code>BEDtools</code> are implemented as <code>dplyr</code> pipes (e.g., see the <a href="#groupby"><code>group_by</code></a> section).</p>
<p>All methods accept one or more sets of <code>x</code> and <code>y</code> intervals, which must either be created using the <a href="#reading%20data"><code>read</code> methods</a>, or have <code>chrom</code>, <code>start</code> and <code>end</code> columns.</p>
<div id="single-set-operations" class="section level2 tabset">
<h2>Single set operations</h2>
<p>These methods operate on a single set of intervals:</p>
<ul><li><code><a href="../reference/bed_sort.html">bed_sort()</a></code>: order intervals</li>
<li><code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>: Cluster (but don&rsquo;t merge) overlapping/nearby intervals.</li>
<li><code><a href="../reference/bed_complement.html">bed_complement()</a></code>: extract intervals <em>not</em> represented by an interval file.</li>
<li><code><a href="../reference/bed_merge.html">bed_merge()</a></code>: combine overlapping and nearby intervals into a single interval.</li>
<li><code><a href="../reference/bed_flank.html">bed_flank()</a></code>: Generate new flanking intervals</li>
<li><code><a href="../reference/bed_slop.html">bed_slop()</a></code>: Expand the size of input intervals</li>
<li><code><a href="../reference/bed_shift.html">bed_shift()</a></code>: Shift the coordinates of an input set, bounded by a genome</li>
</ul><div id="sort" class="section level3">
<h3>Sort</h3>
<p><code>bed_sort</code> orders intervals based on a specification. <code>is_sorted</code> asks whether a tbl is already sorted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_sort.html">bed_sort</a></span>(x)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  1678  2678</span>
<span class="co">#&gt; 2   chr1  4843  5843</span>
<span class="co">#&gt; 3   chr1  7988  8988</span>
<span class="co">#&gt; 4   chr1  8968  9968</span>
<span class="co">#&gt; 5   chr1 10810 11810</span>
<span class="co">#&gt; 6   chr1 11933 12933</span>
<span class="co">#&gt; 7   chr1 13387 14387</span>
<span class="co">#&gt; 8   chr1 24039 25039</span>
<span class="co">#&gt; 9   chr1 37370 38370</span>
<span class="co">#&gt; 10  chr1 39300 40300</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(y)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
<div id="cluster" class="section level3">
<h3>Cluster</h3>
<p><code>bed_cluster</code> identifies clustered intervals based on a distance specification and assigns them a unique <code>.id</code>.</p>
<p><img src="img/functions-cluster_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_cluster.html">bed_cluster</a></span>(x, <span class="dt">max_dist =</span> <span class="dv">1000</span>)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 4</span>
<span class="co">#&gt;    chrom     start       end   .id</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr2 145564847 145565847 47168</span>
<span class="co">#&gt; 2   chr6  92366412  92367412 29861</span>
<span class="co">#&gt; 3   chr2 172970424 172971424 55976</span>
<span class="co">#&gt; 4   chr9 133933926 133934926 43132</span>
<span class="co">#&gt; 5   chr3 128566128 128567128 41599</span>
<span class="co">#&gt; 6   chr2  19315229  19316229  6293</span>
<span class="co">#&gt; 7  chr15  61594140  61595140 20006</span>
<span class="co">#&gt; 8   chr6  36326219  36327219 11536</span>
<span class="co">#&gt; 9  chr11  48007662  48008662 15361</span>
<span class="co">#&gt; 10 chr13  75882345  75883345 24681</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="complement" class="section level3">
<h3>Complement</h3>
<p><code>bed_complement</code> identifies intervals in a genome that are not covered by an input.</p>
<p><img src="img/functions-complement_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human)
<span class="co">#&gt; # A tibble: 723,752 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1     1  2197</span>
<span class="co">#&gt; 2   chr1  4666  6958</span>
<span class="co">#&gt; 3   chr1  7958  8132</span>
<span class="co">#&gt; 4   chr1  9132 10270</span>
<span class="co">#&gt; 5   chr1 11270 17029</span>
<span class="co">#&gt; 6   chr1 18755 19011</span>
<span class="co">#&gt; 7   chr1 20457 28417</span>
<span class="co">#&gt; 8   chr1 29417 30678</span>
<span class="co">#&gt; 9   chr1 32193 37897</span>
<span class="co">#&gt; 10  chr1 38897 43414</span>
<span class="co">#&gt; # ... with 723,742 more rows</span></code></pre></div>
</div>
<div id="merge" class="section level3">
<h3>Merge</h3>
<p><code>bed_merge</code> identifies overlapping intervals and reports new merged ones. <code>is_merged</code> asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.</p>
<p><img src="img/functions-merge_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="fl">1e6</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> n)
<span class="kw"><a href="../reference/is_merged.html">is_merged</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

<span class="co"># add some signal</span>
x &lt;-<span class="st"> </span>x %&gt;%<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">signal =</span> <span class="kw">runif</span>(n))

<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x, <span class="dt">maxs =</span> <span class="kw">max</span>(signal))
<span class="co">#&gt; # A tibble: 723,639 &times; 4</span>
<span class="co">#&gt;    chrom start   end signal</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  2200  3200      .</span>
<span class="co">#&gt; 2   chr1  3374  4859      .</span>
<span class="co">#&gt; 3   chr1  5564  6564      .</span>
<span class="co">#&gt; 4   chr1 10852 11852      .</span>
<span class="co">#&gt; 5   chr1 13761 14761      .</span>
<span class="co">#&gt; 6   chr1 15949 16949      .</span>
<span class="co">#&gt; 7   chr1 17384 18384      .</span>
<span class="co">#&gt; 8   chr1 20168 21168      .</span>
<span class="co">#&gt; 9   chr1 21182 22182      .</span>
<span class="co">#&gt; 10  chr1 24921 25921      .</span>
<span class="co">#&gt; # ... with 723,629 more rows</span></code></pre></div>
</div>
<div id="flank" class="section level3">
<h3>Flank</h3>
<p><code>bed_flank</code> creates new intervals that flank &ndash; but do not contain &ndash; the input intervals.</p>
<p><img src="img/functions-flank_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_flank.html">bed_flank</a></span>(x, genome_human, <span class="dt">both  =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    50   150</span>
<span class="co">#&gt; 2  chr1   125   225</span></code></pre></div>
</div>
<div id="slop" class="section level3">
<h3>Slop</h3>
<p><code>bed_slop</code> pads input intervals based on a specification</p>
<p><img src="img/functions-slop_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_slop.html">bed_slop</a></span>(x, genome_human, <span class="dt">both =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    10   220</span>
<span class="co">#&gt; 2  chr1   125   335</span></code></pre></div>
</div>
<div id="shift" class="section level3">
<h3>Shift</h3>
<p><code>bed_shift</code> adjusts coordinates toward <code>start</code> or <code>end</code> by a defined <code>size</code>. Intervals created out of bounds are removed, or trimmed.</p>
<p><img src="img/functions-shift_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_shift.html">bed_shift</a></span>(x, genome_human, <span class="dt">size =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1   125   150</span>
<span class="co">#&gt; 2  chr1   200   225</span></code></pre></div>
</div>
<div id="interval-spacing" class="section level3">
<h3>Interval spacing</h3>
<p>Interval spacing is easily computed using <code>dplyr</code>. Overlapping intervals must first be merged. Spacing for the first interval of each chromosome is undefined.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x)

x %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">.spacing =</span> start -<span class="st"> </span><span class="kw">lag</span>(end))
<span class="co">#&gt; Source: local data frame [722,910 x 4]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start   end .spacing</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1    98  1098       NA</span>
<span class="co">#&gt; 2   chr1 11559 12559    10461</span>
<span class="co">#&gt; 3   chr1 12654 14014       95</span>
<span class="co">#&gt; 4   chr1 16513 17666     2499</span>
<span class="co">#&gt; 5   chr1 19148 20148     1482</span>
<span class="co">#&gt; 6   chr1 22340 24081     2192</span>
<span class="co">#&gt; 7   chr1 25940 26940     1859</span>
<span class="co">#&gt; 8   chr1 26980 28575       40</span>
<span class="co">#&gt; 9   chr1 29409 30409      834</span>
<span class="co">#&gt; 10  chr1 30418 31418        9</span>
<span class="co">#&gt; # ... with 722,900 more rows</span></code></pre></div>
</div>
</div>
<div id="multiple-set-operations" class="section level2 tabset">
<h2>Multiple set operations</h2>
<p>These methods compare two sets of intervals:</p>
<ul><li><code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>: find overlapping intervals</li>
<li><code><a href="../reference/bed_map.html">bed_map()</a></code>: apply a function to selected columns for overlapping intervals</li>
<li><code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>: Remove intervals based on overlaps between two files</li>
<li><code><a href="../reference/bed_window.html">bed_window()</a></code>: Find overlapping intervals within a window</li>
<li><code><a href="../reference/bed_closest.html">bed_closest()</a></code>: find the closest intervals independent of overlaps</li>
</ul><div id="intersection" class="section level3">
<h3>Intersection</h3>
<p><code>bed_intersect</code> is implemented using an <a href="https://github.com/ekg/intervaltree/">interval tree</a> in <code>Rcpp</code>. Column names in the result have <code>.x</code> and <code>.y</code> suffixes, and an <code>.overlap</code> column contains the size of the intersection (values of <code>0</code> indicate book-ended, or touching intervals). See the <a href="#benchmarks">benchmarks</a> section for timing. Though <code>bed_intersect</code> is pretty fast already, we intend to further improve upon this by parallization with <a href="https://rcppcore.github.io/RcppParallel/"><code>RcppParallel</code></a>.</p>
<p><img src="img/functions-intersect_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># intersect two sets of 1e6 intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 645,672 &times; 6</span>
<span class="co">#&gt;    chrom start.x end.x start.y end.y .overlap</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt; &lt;int&gt;   &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   11505 12505   11236 12236      731</span>
<span class="co">#&gt; 2   chr1   11505 12505   12386 13386      119</span>
<span class="co">#&gt; 3   chr1   12600 13600   12386 13386      786</span>
<span class="co">#&gt; 4   chr1   12600 13600   13175 14175      425</span>
<span class="co">#&gt; 5   chr1   12600 13600   13353 14353      247</span>
<span class="co">#&gt; 6   chr1   14866 15866   14341 15341      475</span>
<span class="co">#&gt; 7   chr1   18077 19077   18738 19738      339</span>
<span class="co">#&gt; 8   chr1   19833 20833   19735 20735      902</span>
<span class="co">#&gt; 9   chr1   21305 22305   21189 22189      884</span>
<span class="co">#&gt; 10  chr1   21305 22305   22161 23161      144</span>
<span class="co">#&gt; # ... with 645,662 more rows</span>

<span class="co"># A records with no overlaps (i.e., `-v`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y, <span class="dt">invert =</span> <span class="ot">TRUE</span>)
<span class="co">#&gt; # A tibble: 524,272 &times; 3</span>
<span class="co">#&gt;    chrom    start      end</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chrY 59361797 59362797</span>
<span class="co">#&gt; 2   chrY 59360628 59361628</span>
<span class="co">#&gt; 3   chrY 59337921 59338921</span>
<span class="co">#&gt; 4   chrY 59323831 59324831</span>
<span class="co">#&gt; 5   chrY 59321564 59322564</span>
<span class="co">#&gt; 6   chrY 59319849 59320849</span>
<span class="co">#&gt; 7   chrY 59317664 59318664</span>
<span class="co">#&gt; 8   chrY 59308803 59309803</span>
<span class="co">#&gt; 9   chrY 59293051 59294051</span>
<span class="co">#&gt; 10  chrY 59288344 59289344</span>
<span class="co">#&gt; # ... with 524,262 more rows</span></code></pre></div>
<p>One can achieve behaviour similar to BEDtools by combining <code>bed_intersect</code> with <code>dplyr</code> tools.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># `x` records with overlaps (i.e., `-wa`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x)
<span class="co">#&gt; # A tibble: 645,672 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 11505 12505</span>
<span class="co">#&gt; 2   chr1 11505 12505</span>
<span class="co">#&gt; 3   chr1 12600 13600</span>
<span class="co">#&gt; 4   chr1 12600 13600</span>
<span class="co">#&gt; 5   chr1 12600 13600</span>
<span class="co">#&gt; 6   chr1 14866 15866</span>
<span class="co">#&gt; 7   chr1 18077 19077</span>
<span class="co">#&gt; 8   chr1 19833 20833</span>
<span class="co">#&gt; 9   chr1 21305 22305</span>
<span class="co">#&gt; 10  chr1 21305 22305</span>
<span class="co">#&gt; # ... with 645,662 more rows</span>

<span class="co"># `y` records with overlaps (i.e., `-wb`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.y, <span class="dt">end =</span> end.y)
<span class="co">#&gt; # A tibble: 645,672 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 11236 12236</span>
<span class="co">#&gt; 2   chr1 12386 13386</span>
<span class="co">#&gt; 3   chr1 12386 13386</span>
<span class="co">#&gt; 4   chr1 13175 14175</span>
<span class="co">#&gt; 5   chr1 13353 14353</span>
<span class="co">#&gt; 6   chr1 14341 15341</span>
<span class="co">#&gt; 7   chr1 18738 19738</span>
<span class="co">#&gt; 8   chr1 19735 20735</span>
<span class="co">#&gt; 9   chr1 21189 22189</span>
<span class="co">#&gt; 10  chr1 22161 23161</span>
<span class="co">#&gt; # ... with 645,662 more rows</span>

<span class="co"># Unique records in `x` (i.e., `-u`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x) %&gt;%<span class="st"> </span><span class="kw">unique</span>()
<span class="co">#&gt; # A tibble: 475,651 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 11505 12505</span>
<span class="co">#&gt; 2   chr1 12600 13600</span>
<span class="co">#&gt; 3   chr1 14866 15866</span>
<span class="co">#&gt; 4   chr1 18077 19077</span>
<span class="co">#&gt; 5   chr1 19833 20833</span>
<span class="co">#&gt; 6   chr1 21305 22305</span>
<span class="co">#&gt; 7   chr1 21450 22450</span>
<span class="co">#&gt; 8   chr1 23465 24465</span>
<span class="co">#&gt; 9   chr1 36727 37727</span>
<span class="co">#&gt; 10  chr1 44511 45511</span>
<span class="co">#&gt; # ... with 475,641 more rows</span>

<span class="co"># Count `y` intervals that intersect each `x` interval</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">group_by</span>(chrom, start.x, end.x) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">count =</span> <span class="kw">n</span>())
<span class="co">#&gt; Source: local data frame [475,651 x 4]</span>
<span class="co">#&gt; Groups: chrom, start.x [?]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start.x end.x count</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   11505 12505     2</span>
<span class="co">#&gt; 2   chr1   12600 13600     3</span>
<span class="co">#&gt; 3   chr1   14866 15866     1</span>
<span class="co">#&gt; 4   chr1   18077 19077     1</span>
<span class="co">#&gt; 5   chr1   19833 20833     1</span>
<span class="co">#&gt; 6   chr1   21305 22305     2</span>
<span class="co">#&gt; 7   chr1   21450 22450     2</span>
<span class="co">#&gt; 8   chr1   23465 24465     1</span>
<span class="co">#&gt; 9   chr1   36727 37727     1</span>
<span class="co">#&gt; 10  chr1   44511 45511     1</span>
<span class="co">#&gt; # ... with 475,641 more rows</span></code></pre></div>
</div>
<div id="map" class="section level3">
<h3>Map</h3>
<p><code>bed_map</code> maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs.</p>
<p><img src="img/functions-map_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bedfile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)
bgfile  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)

x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(bgfile)

<span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">means =</span> <span class="kw">mean</span>(value), <span class="dt">sds =</span> <span class="kw">sd</span>(value))
<span class="co">#&gt; # A tibble: 730 &times; 8</span>
<span class="co">#&gt;    chrom    start      end    means      sds     name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr22 16150259 16193004 7.914286 7.570731     &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 2  chr22 16162065 16172265 1.000000       NA     &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 3  chr22 16179617 16181004       NA       NA BC017398     1      -</span>
<span class="co">#&gt; 4  chr22 16239287 16239327       NA       NA DQ590589     1      +</span>
<span class="co">#&gt; 5  chr22 16240245 16240277       NA       NA DQ573684     1      -</span>
<span class="co">#&gt; 6  chr22 16240300 16240340       NA       NA DQ595048     1      -</span>
<span class="co">#&gt; 7  chr22 16241085 16241125       NA       NA DQ590589     1      +</span>
<span class="co">#&gt; 8  chr22 16242753 16242785       NA       NA DQ573684     1      -</span>
<span class="co">#&gt; 9  chr22 16243382 16243414       NA       NA DQ573684     1      -</span>
<span class="co">#&gt; 10 chr22 16243447 16243487       NA       NA DQ587539     1      -</span>
<span class="co">#&gt; # ... with 720 more rows</span></code></pre></div>
</div>
<div id="subtract" class="section level3">
<h3>Subtract</h3>
<p><code>bed_substract()</code> removes <code>x</code> intervals that intersect with <code>y</code>.</p>
<p><img src="img/functions-subtract_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y)
<span class="co">#&gt; # A tibble: 957,376 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  1493  2493</span>
<span class="co">#&gt; 2   chr1  3102  3490</span>
<span class="co">#&gt; 3   chr1  7915  8915</span>
<span class="co">#&gt; 4   chr1 11941 12941</span>
<span class="co">#&gt; 5   chr1 16236 16603</span>
<span class="co">#&gt; 6   chr1 17770 18172</span>
<span class="co">#&gt; 7   chr1 23823 24823</span>
<span class="co">#&gt; 8   chr1 25820 26820</span>
<span class="co">#&gt; 9   chr1 49836 50836</span>
<span class="co">#&gt; 10  chr1 57740 58740</span>
<span class="co">#&gt; # ... with 957,366 more rows</span></code></pre></div>
</div>
<div id="window" class="section level3">
<h3>Window</h3>
<p><code><a href="../reference/bed_window.html">bed_window()</a></code> identifies <code>y</code> intervals that intersect an expanded window of <code>x</code> intervals.</p>
<p><img src="img/functions-window_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="co"># a few intersections</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span>

<span class="co"># can be expanded by casting a wider net</span>
<span class="kw"><a href="../reference/bed_window.html">bed_window</a></span>(x, y, genome, <span class="dt">both =</span> <span class="fl">1e6</span>)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span></code></pre></div>
</div>
<div id="closest" class="section level3">
<h3>Closest</h3>
<p><code><a href="../reference/bed_closest.html">bed_closest()</a></code> identifies <code>y</code> intervals that are closest to <code>x</code>.</p>
<p><img src="img/functions-closest_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y)
<span class="co">#&gt; # A tibble: 94 &times; 7</span>
<span class="co">#&gt;    chrom   start.x     end.x   start.y     end.y .overlap .distance</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;    &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   9652081   9653081   8431546   8432546        0  -1219535</span>
<span class="co">#&gt; 2   chr1  39712640  39713640  35648354  35649354        0  -4063286</span>
<span class="co">#&gt; 3   chr1  57183909  57184909  64507324  64508324        0   7322415</span>
<span class="co">#&gt; 4   chr1  75160036  75161036  64507324  64508324        0 -10651712</span>
<span class="co">#&gt; 5   chr1 174684091 174685091 176463711 176464711        0   1778620</span>
<span class="co">#&gt; 6   chr1 191689881 191690881 176463711 176464711        0 -15225170</span>
<span class="co">#&gt; 7   chr1 240614402 240615402 244198901 244199901        0   3583499</span>
<span class="co">#&gt; 8   chr1 244134217 244135217 244198901 244199901        0     63684</span>
<span class="co">#&gt; 9   chr1 247812533 247813533 244198901 244199901        0  -3612632</span>
<span class="co">#&gt; 10 chr10  14455428  14456428 107906686 107907686        0  93450258</span>
<span class="co">#&gt; # ... with 84 more rows</span></code></pre></div>
</div>
</div>
<div id="randomzing-intervals" class="section level2 tabset">
<h2>Randomzing intervals</h2>
<p><code>valr</code> provides methods for creating new random intervals or permutations of existing intervals:</p>
<ul><li><code>bed_random</code> generates random intervals from an input <code>genome</code>.</li>
<li><code>bed_shuffle</code> shuffles coordinates given a set of input intervals.</li>
<li>Random sampling of input intervals is done with <code>dplyr</code>.</li>
</ul><div id="random" class="section level3">
<h3>Random</h3>
<p><code>bed_random</code> generates random intervals from an input genome. The numbers of intervals from each <code>chrom</code> are proporitional to each chrom <code>size</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="fl">1e6</span>, <span class="dt">length =</span> <span class="fl">1e3</span>)
x
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  48647369  48648369</span>
<span class="co">#&gt; 2  chr17  10916412  10917412</span>
<span class="co">#&gt; 3   chr4 139316062 139317062</span>
<span class="co">#&gt; 4   chr3  37230847  37231847</span>
<span class="co">#&gt; 5   chr9  99007887  99008887</span>
<span class="co">#&gt; 6   chrX  44502251  44503251</span>
<span class="co">#&gt; 7   chrX 108017266 108018266</span>
<span class="co">#&gt; 8  chr13  12246522  12247522</span>
<span class="co">#&gt; 9  chr20  23295502  23296502</span>
<span class="co">#&gt; 10  chr8  89602315  89603315</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="co"># numbers of sampled intervals are proportional to chrom size</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">n =</span> <span class="kw">n</span>()) %&gt;%<span class="st"> </span><span class="kw">arrange</span>(<span class="kw">desc</span>(n))
<span class="co">#&gt; # A tibble: 25 &times; 2</span>
<span class="co">#&gt;    chrom     n</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 80293</span>
<span class="co">#&gt; 2   chr2 78694</span>
<span class="co">#&gt; 3   chr3 63852</span>
<span class="co">#&gt; 4   chr4 61813</span>
<span class="co">#&gt; 5   chr5 58578</span>
<span class="co">#&gt; 6   chr6 55063</span>
<span class="co">#&gt; 7   chr7 51314</span>
<span class="co">#&gt; 8   chrX 50062</span>
<span class="co">#&gt; 9   chr8 47569</span>
<span class="co">#&gt; 10  chr9 45903</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="sample" class="section level3">
<h3>Sample</h3>
<p>Sampling can be done using <code>dplyr</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="co"># sample by number</span>
<span class="kw">sample_n</span>(x, <span class="fl">1e3</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 1,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr12  36927247  36928247</span>
<span class="co">#&gt; 2   chr1 209490106 209491106</span>
<span class="co">#&gt; 3   chr6 111017565 111018565</span>
<span class="co">#&gt; 4  chr13  87195146  87196146</span>
<span class="co">#&gt; 5  chr21  17935961  17936961</span>
<span class="co">#&gt; 6   chr5 138769851 138770851</span>
<span class="co">#&gt; 7   chrX  79640300  79641300</span>
<span class="co">#&gt; 8  chr12 124597747 124598747</span>
<span class="co">#&gt; 9   chr2 141601234 141602234</span>
<span class="co">#&gt; 10 chr20  52916739  52917739</span>
<span class="co">#&gt; # ... with 990 more rows</span>

<span class="co"># or fraction</span>
<span class="kw">sample_frac</span>(x, <span class="fl">0.1</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 100,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr14 101477058 101478058</span>
<span class="co">#&gt; 2  chr13  53925464  53926464</span>
<span class="co">#&gt; 3  chr19  39835509  39836509</span>
<span class="co">#&gt; 4  chr14  29040944  29041944</span>
<span class="co">#&gt; 5   chr1 178548053 178549053</span>
<span class="co">#&gt; 6   chr1 225574178 225575178</span>
<span class="co">#&gt; 7   chr3  73349408  73350408</span>
<span class="co">#&gt; 8   chr9 108174203 108175203</span>
<span class="co">#&gt; 9   chrX 135319922 135320922</span>
<span class="co">#&gt; 10  chrX 141686808 141687808</span>
<span class="co">#&gt; # ... with 99,990 more rows</span>

<span class="co"># or sample intervals within groups</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">sample_n</span>(<span class="dv">1</span>)
<span class="co">#&gt; Source: local data frame [25 x 3]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 142381530 142382530</span>
<span class="co">#&gt; 2  chr10 126375563 126376563</span>
<span class="co">#&gt; 3  chr11  16957854  16958854</span>
<span class="co">#&gt; 4  chr12 126412153 126413153</span>
<span class="co">#&gt; 5  chr13  99578724  99579724</span>
<span class="co">#&gt; 6  chr14  50277927  50278927</span>
<span class="co">#&gt; 7  chr15  74742996  74743996</span>
<span class="co">#&gt; 8  chr16    242305    243305</span>
<span class="co">#&gt; 9  chr17  72480369  72481369</span>
<span class="co">#&gt; 10 chr18  50493329  50494329</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="shuffle" class="section level3">
<h3>Shuffle</h3>
<p><code>bed_shuffle</code> shuffles input intervals. Interval sizes are equal in the input and output.</p>
<pre><code>#&gt; Warning in bind_rows_(x, .id): binding factor and character vector,
#&gt; coercing into character vector</code></pre>
<p><img src="img/functions-shuffle_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human)
y
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;   &lt;fctr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr3  64090381  64090401</span>
<span class="co">#&gt; 2   chr3 149879636 149879646</span>

<span class="kw">any</span>(x$start ==<span class="st"> </span>y$start)
<span class="co">#&gt; [1] FALSE</span>

<span class="kw">all</span>(x$end -<span class="st"> </span>x$start ==<span class="st"> </span>y$end -<span class="st"> </span>y$start)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
</div>
<div id="other-topics" class="section level2">
<h2>Other topics</h2>
<div id="grouping-related-intervals" class="section level3">
<h3>Grouping related intervals</h3>
<p>The <code>group_by</code> operation from <code>dplyr</code> serves a similar purpose to BEDtools <code>groupby</code>, and can be combined with <code>dplyr::summarize</code> to calculate summary statistics from groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">overlap.total =</span> <span class="kw">sum</span>(.overlap))
<span class="co">#&gt; # A tibble: 24 &times; 2</span>
<span class="co">#&gt;    chrom overlap.total</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1      26177318</span>
<span class="co">#&gt; 2  chr10      14101570</span>
<span class="co">#&gt; 3  chr11      13873481</span>
<span class="co">#&gt; 4  chr12      14212163</span>
<span class="co">#&gt; 5  chr13      11955316</span>
<span class="co">#&gt; 6  chr14      11161495</span>
<span class="co">#&gt; 7  chr15      10802296</span>
<span class="co">#&gt; 8  chr16       9259617</span>
<span class="co">#&gt; 9  chr17       8341247</span>
<span class="co">#&gt; 10 chr18       8203401</span>
<span class="co">#&gt; # ... with 14 more rows</span></code></pre></div>
</div>
</div>
</div>
<div id="interval-statistics" class="section level1 tabset">
<h1>Interval Statistics</h1>
<p><code>valr</code> provides several methods to assess statistical properties of interval sets including:</p>
<ul><li><code><a href="../reference/bed_fisher.html">bed_fisher()</a></code>: measure overlap significance of two sets of intervals.</li>
<li><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>: quantify absolute distance between query intervals and closest reference intervals</li>
<li><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code>: quantify relative distances between query intervals and closest reference intervals</li>
<li><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>: quantify extent of overlap between two sets of intervals</li>
</ul><p>Several of these methods were described in the <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">Genometricorr</a> software package.</p>
<div id="fishers-test" class="section level2">
<h2>Fisher&rsquo;s test</h2>
<p>The Fisher&rsquo;s test assesses whether two sets of intervals are drawn from the same background genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_fisher.html">bed_fisher</a></span>(x, y, genome_human)
<span class="co">#&gt;   estimate p.value conf.low conf.high                             method</span>
<span class="co">#&gt; 1 2.366201       0 2.352076  2.380424 Fisher's Exact Test for Count Data</span>
<span class="co">#&gt;   alternative</span>
<span class="co">#&gt; 1   two.sided</span></code></pre></div>
</div>
<div id="absolute-distance" class="section level2">
<h2>Absolute distance</h2>
<p><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code> computes the absolute distance between the midpoint of query intervals and the closest midpoints of a set of reference intervals. Absolute distances are scaled by the inter-reference gap for the chromosome as follows. For <code>Q</code> total query points and <code>R</code> reference points on a chromosome, scale the distance for each query point <code>i</code> to the closest reference point by the inter-reference gap for each chromosome.</p>
<p><span class="math display">\[
d_i(x,y) = min_{\substack{k}}(|q_i - r_k|)\frac{R}{Length\ of\ chromsome}
\]</span></p>
<p>By default both absolute and scaled distances are reported as <code>absdist</code> and <code>scaled_absdist</code> respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_absdist.html">bed_absdist</a></span>(x, y, genome_human)
<span class="co">#&gt; # A tibble: 1,000,000 &times; 5</span>
<span class="co">#&gt;    chrom start   end absdist scaled_absdist</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;   &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1   chr1  2689  3689    2660     0.85831610</span>
<span class="co">#&gt; 2   chr1  6206  7206     406     0.13100614</span>
<span class="co">#&gt; 3   chr1  9158 10158     428     0.13810500</span>
<span class="co">#&gt; 4   chr1  9412 10412     174     0.05614549</span>
<span class="co">#&gt; 5   chr1  9931 10931     345     0.11132295</span>
<span class="co">#&gt; 6   chr1 10701 11701     396     0.12777939</span>
<span class="co">#&gt; 7   chr1 13679 14679      80     0.02581402</span>
<span class="co">#&gt; 8   chr1 15499 16499     974     0.31428567</span>
<span class="co">#&gt; 9   chr1 16297 17297    1772     0.57178050</span>
<span class="co">#&gt; 10  chr1 16361 17361    1836     0.59243171</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="relative-distance" class="section level2">
<h2>Relative distance</h2>
<p><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> computes the relative distance between a query interval and the two closest reference intervals. The relative distance can range between <code>[0,0.5]</code>. If the overall distribution of distances is biased toward zero, then the query and reference intervals tend to be closer in space (<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">17</a>). By default a summary with the frequency of each relative distance is reported.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_reldist.html">bed_reldist</a></span>(x, y)
<span class="co">#&gt; # A tibble: 51 &times; 4</span>
<span class="co">#&gt;    reldist counts  total       freq</span>
<span class="co">#&gt;      &lt;dbl&gt;  &lt;int&gt;  &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1     0.00  19690 999949 0.01969100</span>
<span class="co">#&gt; 2     0.01  20141 999949 0.02014203</span>
<span class="co">#&gt; 3     0.02  19973 999949 0.01997402</span>
<span class="co">#&gt; 4     0.03  19912 999949 0.01991302</span>
<span class="co">#&gt; 5     0.04  19838 999949 0.01983901</span>
<span class="co">#&gt; 6     0.05  20051 999949 0.02005202</span>
<span class="co">#&gt; 7     0.06  20067 999949 0.02006802</span>
<span class="co">#&gt; 8     0.07  19977 999949 0.01997802</span>
<span class="co">#&gt; 9     0.08  20024 999949 0.02002502</span>
<span class="co">#&gt; 10    0.09  19957 999949 0.01995802</span>
<span class="co">#&gt; # ... with 41 more rows</span></code></pre></div>
</div>
<div id="jaccard-similarity" class="section level2">
<h2>Jaccard similarity</h2>
<p><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code> quantifies the extent of overlap between to sets of intervals. The Jaccard statistic takes values of <code>[0,1]</code> and is measured as:</p>
<p><span class="math display">\[
J(x,y) = \frac{\mid x \bigcap y \mid}
              {\mid x \bigcup y \mid} 
       = 
         \frac{\mid x \bigcap y \mid}
              {\mid x \mid + \mid y \mid - \mid x \bigcap y \mid}
\]</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_jaccard.html">bed_jaccard</a></span>(x, y)
<span class="co">#&gt; # A tibble: 1 &times; 4</span>
<span class="co">#&gt;       len_i      len_u   jaccard      n</span>
<span class="co">#&gt;       &lt;int&gt;      &lt;int&gt;     &lt;dbl&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1 322504062 2000000000 0.1922533 644868</span></code></pre></div>
</div>
</div>
<div id="benchmarks" class="section level1">
<h1>Benchmarks</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># two sets of 1e6 random 1 kb intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw">library</span>(microbenchmark)
<span class="kw">microbenchmark</span>(
  <span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human),
  <span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x),
  <span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human),
  <span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human),
  <span class="dt">times =</span> <span class="dv">1</span>,
  <span class="dt">unit =</span> <span class="st">'s'</span>
)
<span class="co">#&gt; Unit: seconds</span>
<span class="co">#&gt;                             expr       min        lq      mean    median</span>
<span class="co">#&gt;         bed_random(genome_human) 0.1166527 0.1166527 0.1166527 0.1166527</span>
<span class="co">#&gt;                bed_closest(x, y) 5.6083553 5.6083553 5.6083553 5.6083553</span>
<span class="co">#&gt;              bed_intersect(x, y) 2.0672902 2.0672902 2.0672902 2.0672902</span>
<span class="co">#&gt;                     bed_merge(x) 3.1950880 3.1950880 3.1950880 3.1950880</span>
<span class="co">#&gt;               bed_subtract(x, y) 3.9432268 3.9432268 3.9432268 3.9432268</span>
<span class="co">#&gt;  bed_complement(x, genome_human) 3.6628542 3.6628542 3.6628542 3.6628542</span>
<span class="co">#&gt;     bed_shuffle(x, genome_human) 1.1182704 1.1182704 1.1182704 1.1182704</span>
<span class="co">#&gt;         uq       max neval</span>
<span class="co">#&gt;  0.1166527 0.1166527     1</span>
<span class="co">#&gt;  5.6083553 5.6083553     1</span>
<span class="co">#&gt;  2.0672902 2.0672902     1</span>
<span class="co">#&gt;  3.1950880 3.1950880     1</span>
<span class="co">#&gt;  3.9432268 3.9432268     1</span>
<span class="co">#&gt;  3.6628542 3.6628542     1</span>
<span class="co">#&gt;  1.1182704 1.1182704     1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#overview">Overview</a><ul class="nav nav-pills nav-stacked"><li><a href="#installation">Installation</a></li>
      <li><a href="#comparison-to-other-tools">Comparison to other tools</a></li>
      <li><a href="#non-standard-evaluation">Non-standard evaluation</a></li>
      <li><a href="#getting-started">Getting started</a></li>
      </ul></li>
      <li><a href="#reading-data">Reading data</a><ul class="nav nav-pills nav-stacked"><li><a href="#bed-files">BED files</a></li>
      <li><a href="#genome-files">Genome files</a></li>
      <li><a href="#bedgraph-files">bedGraph files</a></li>
      <li><a href="#vcf-files">VCF files</a></li>
      </ul></li>
      <li><a href="#interval-manipulations">Interval manipulations</a><ul class="nav nav-pills nav-stacked"><li><a href="#single-set-operations">Single set operations</a></li>
      <li><a href="#multiple-set-operations">Multiple set operations</a></li>
      <li><a href="#randomzing-intervals">Randomzing intervals</a></li>
      <li><a href="#other-topics">Other topics</a></li>
      </ul></li>
      <li><a href="#interval-statistics">Interval Statistics</a><ul class="nav nav-pills nav-stacked"><li><a href="#fishers-test">Fisher&rsquo;s test</a></li>
      <li><a href="#absolute-distance">Absolute distance</a></li>
      <li><a href="#relative-distance">Relative distance</a></li>
      <li><a href="#jaccard-similarity">Jaccard similarity</a></li>
      </ul></li>
      <li><a href="#benchmarks">Benchmarks</a></li>
      </ul></div>
      </div>

</div>


      <footer><p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer></div>

  </body></html>
