<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title><code>valr</code>: Genome interval arithmetic in R. valr</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container">
      <header><div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/valr-overview.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1><code>valr</code>: Genome interval arithmetic in R</h1>
                        <h4 class="author">Jay Hesselberth <a href="mailto:jay.hesselberth@gmail.com">jay.hesselberth@gmail.com</a></h4>
            
            <h4 class="date">2016-10-12</h4>
          </div>

    
    
<div id="overview" class="section level1">
<h1>Overview</h1>
<p><code>valr</code> provides tools to read and manipulate genome intervals and signals, similar to the standalone <a href="http://bedtools.readthedocs.org/en/latest/"><code>BEDtools</code></a> suite. <code>valr</code> enables BEDtools-like analysis in the R/RStudio environment, and uses modern R tools for a terse, expressive syntax. Compute-intensive algorithms are implemented in <a href="http://www.rcpp.org/"><code>Rcpp</code></a>/C++, and many methods take advantage of the speed and grouping capability provided by <a href="https://github.com/hadley/dplyr"><code>dplyr</code></a>.</p>
<p>We provide several introductions to <code>valr</code>:</p>
<ul><li>This comprehensive <code>vignette</code> covering the core methods.</li>
<li>A tutorial that demonstrates how to use <code>valr</code> in &ldquo;real-world&rdquo; applications.</li>
<li>A <code>shiny</code> application that demonstrates interactive analysis of genome-scale data sets.</li>
</ul><div id="installation" class="section level2">
<h2>Installation</h2>
<p><code>valr</code> can be installed from github, and will be eventually deposited in CRAN.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">'jayhesselberth/valr'</span>)</code></pre></div>
</div>
<div id="comparison-to-other-tools" class="section level2">
<h2>Comparison to other tools</h2>
<p><strong>Why another tool set for interval manipulations?</strong> We already have <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a>, <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>, <a href="https://pythonhosted.org/pybedtools/">pybedtools</a>, <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://cran.r-project.org/web/packages/bedr/index.html">bedr</a> and <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a>.</p>
<p>We were motivated to develop a toolset that:</p>
<ul><li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the &ldquo;modern R&rdquo; syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code>%&gt;%</code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visulaizations with <a href="http://shiny.rstudio.com/"><code>shiny</code></a>.</li>
</ul><p>We anticipate <code>valr</code> will mainly be used for analysis of pre-processed data in BED, bedGraph and VCF formats. Most users will have processed their aligned reads from BAM format to bedGraph, so we do not foresee supporting BAM directly. We would entertain requests for GTF / GFF support if there is interest, as <code>tidyr</code> makes it easy to convert these to BED12.</p>
<p>Certain algorithms in <code>valr</code> were implemented in <code>Rcpp</code> (including <a href="https://github.com/jayhesselberth/valr/tree/master/src">intersect, merge, subtract, closest</a>) to enable fluid interactive analysis. See the <a href="#benchmarks">benchmarks</a> section for details.</p>
</div>
<div id="non-standard-evaluation" class="section level2">
<h2>Non-standard evaluation</h2>
<p>Several of the methods in <code>valr</code> use <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html">NSE</a> for an expressive syntax. Columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">mean =</span> <span class="kw">mean</span>(value), <span class="dt">var =</span> <span class="kw">var</span>(value))
<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x, <span class="dt">concat =</span> <span class="kw"><a href="../reference/bed_map.html">concat</a></span>(value), <span class="dt">max =</span> <span class="kw">max</span>(value))</code></pre></div>
</div>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Here is an example using <code>valr</code> that creates</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)</code></pre></div>
</div>
</div>
<div id="reading-data" class="section level1 tabset">
<h1>Reading data</h1>
<p><code>valr</code> has several methods to read interval data. These methods:</p>
<ul><li>Take local files or URLs as input.</li>
<li>Return a <code>data.frame</code> in <code>tibble::tbl_df</code> format.</li>
<li>Assign consistent <code>chrom</code>, <code>start</code> and <code>end</code> column names.</li>
<li>Use <a href="https://github.com/hadley/readr"><code>readr</code></a> for speed.</li>
<li>Coerce column types.</li>
</ul><p>The methods include:</p>
<ul><li><code><a href="../reference/read_bed.html">read_bed()</a></code>: read a BED3+ file</li>
<li><code><a href="../reference/read_bed.html">read_bed12()</a></code>: read a BED12 file</li>
<li><code><a href="../reference/read_bed.html">read_bedgraph()</a></code>: read a bedGraph file</li>
<li><code><a href="../reference/read_genome.html">read_genome()</a></code>: read a UCSC &ldquo;chrom size&rdquo; file</li>
<li><code><a href="../reference/read_vcf.html">read_vcf()</a></code>: read the Variant Call Format</li>
<li><code><a href="../reference/read_bed.html">read_narrowpeak()</a></code>: read narrowPeak files</li>
<li><code><a href="../reference/read_bed.html">read_broadpeak()</a></code>: read broadPeak files</li>
</ul><div id="bed-files" class="section level2">
<h2>BED files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'3fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 3</span>
<span class="co">#&gt;    chrom  start    end</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409</span>
<span class="co">#&gt; 2   chr1  14361  19759</span>
<span class="co">#&gt; 3   chr1  14406  29370</span>
<span class="co">#&gt; 4   chr1  34610  36081</span>
<span class="co">#&gt; 5   chr1  69090  70008</span>
<span class="co">#&gt; 6   chr1 134772 140566</span>
<span class="co">#&gt; 7   chr1 321083 321115</span>
<span class="co">#&gt; 8   chr1 321145 321207</span>
<span class="co">#&gt; 9   chr1 322036 326938</span>
<span class="co">#&gt; 10  chr1 327545 328439</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="dt">n_fields =</span> <span class="dv">6</span>, <span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'6fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409      DDX11L1     3      +</span>
<span class="co">#&gt; 2   chr1  14361  19759       WASH7P    10      -</span>
<span class="co">#&gt; 3   chr1  14406  29370       WASH7P     7      -</span>
<span class="co">#&gt; 4   chr1  34610  36081      FAM138F     3      -</span>
<span class="co">#&gt; 5   chr1  69090  70008        OR4F5     1      +</span>
<span class="co">#&gt; 6   chr1 134772 140566    LOC729737     3      -</span>
<span class="co">#&gt; 7   chr1 321083 321115     DQ597235     1      +</span>
<span class="co">#&gt; 8   chr1 321145 321207     DQ599768     1      +</span>
<span class="co">#&gt; 9   chr1 322036 326938 LOC100133331     3      +</span>
<span class="co">#&gt; 10  chr1 327545 328439    LOC388312     1      +</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed12</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'12fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 3 &times; 12</span>
<span class="co">#&gt;   chrom   start     end      name score strand cds_start cds_end item_rgb</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;   &lt;int&gt;    &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1 4797973 4836816  testgene     1      +   4797973 4836816        .</span>
<span class="co">#&gt; 2 chr10 4848118 4880877 diffchrom     1      +   4848118 4880877        .</span>
<span class="co">#&gt; 3 chr20 5073253 5152630 negstrand     1      -   5073253 5152630        .</span>
<span class="co">#&gt; # ... with 3 more variables: exon_count &lt;int&gt;, exon_sizes &lt;chr&gt;,</span>
<span class="co">#&gt; #   exon_starts &lt;chr&gt;</span></code></pre></div>
</div>
<div id="genome-files" class="section level2">
<h2>Genome files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome_human &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>))</code></pre></div>
</div>
<div id="bedgraph-files" class="section level2">
<h2>bedGraph files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.bg.gz'</span>))
<span class="co">#&gt; # A tibble: 4 &times; 4</span>
<span class="co">#&gt;   chrom    start      end value</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr19 49302000 49302300 -1.00</span>
<span class="co">#&gt; 2 chr19 49302300 49302600 -0.75</span>
<span class="co">#&gt; 3 chr19 49302600 49302900 -0.50</span>
<span class="co">#&gt; 4 chr19 49302900 49303200 -0.25</span></code></pre></div>
</div>
<div id="vcf-files" class="section level2">
<h2>VCF files</h2>
<p><code><a href="../reference/read_vcf.html">read_vcf()</a></code> reads VCF files and assigns <code>chrom</code>, <code>start</code> and <code>end</code> columns to be used to downstream interval comparisons. Note the interval size is calculated as the length of the <code>REF</code> field in the original file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_vcf.html">read_vcf</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.vcf.gz'</span>))
<span class="co">#&gt; # A tibble: 11 &times; 18</span>
<span class="co">#&gt;    CHROM   POS      ID   REF   ALT  QUAL FILTER</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1      1    10    1:10     A     T   100   PASS</span>
<span class="co">#&gt; 2      1    20    1:20     G     C   100   PASS</span>
<span class="co">#&gt; 3      1    30    1:30     C     A   100   PASS</span>
<span class="co">#&gt; 4      1    40    1:40     A     C   100   PASS</span>
<span class="co">#&gt; 5      1 10000 1:10000     G     C   100   PASS</span>
<span class="co">#&gt; 6      1 20000 1:20000     T     A   100   PASS</span>
<span class="co">#&gt; 7      4  5000  4:5000     A     T   100   PASS</span>
<span class="co">#&gt; 8      4  6000  4:6000     C     T   100   PASS</span>
<span class="co">#&gt; 9      X   800   X:800     A     C   100   PASS</span>
<span class="co">#&gt; 10     X   900   X:900     A     T   100   PASS</span>
<span class="co">#&gt; 11     X  1000  X:1000     T     G   100   PASS</span>
<span class="co">#&gt; # ... with 11 more variables: INFO &lt;chr&gt;, FORMAT &lt;chr&gt;, X1 &lt;chr&gt;,</span>
<span class="co">#&gt; #   X2 &lt;chr&gt;, X3 &lt;chr&gt;, X4 &lt;chr&gt;, X5 &lt;chr&gt;, X6 &lt;chr&gt;, chrom &lt;chr&gt;,</span>
<span class="co">#&gt; #   start &lt;int&gt;, end &lt;int&gt;</span></code></pre></div>
</div>
</div>
<div id="interval-manipulations" class="section level1">
<h1>Interval manipulations</h1>
<p><code>valr</code> implements several methods for manipulating sets of intervals. Some methods operate on a single set of intervals, while others compare two sets of intervals.</p>
<p>Many methods the same name as the corresponding <code>BEDtool</code>, and some commonly used <code>BEDtools</code> are implemented as <code>dplyr</code> pipes (e.g., see the <a href="#groupby"><code>group_by</code></a> section).</p>
<p>All methods accept one or more sets of <code>x</code> and <code>y</code> intervals, which must either be created using the <a href="#reading%20data"><code>read</code> methods</a>, or have <code>chrom</code>, <code>start</code> and <code>end</code> columns.</p>
<div id="single-set-operations" class="section level2 tabset">
<h2>Single set operations</h2>
<p>These methods operate on a single set of intervals:</p>
<ul><li><code><a href="../reference/bed_sort.html">bed_sort()</a></code>: order intervals</li>
<li><code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>: Cluster (but don&rsquo;t merge) overlapping/nearby intervals.</li>
<li><code><a href="../reference/bed_complement.html">bed_complement()</a></code>: extract intervals <em>not</em> represented by an interval file.</li>
<li><code><a href="../reference/bed_merge.html">bed_merge()</a></code>: combine overlapping and nearby intervals into a single interval.</li>
<li><code><a href="../reference/bed_flank.html">bed_flank()</a></code>: Generate new flanking intervals</li>
<li><code><a href="../reference/bed_slop.html">bed_slop()</a></code>: Expand the size of input intervals</li>
<li><code><a href="../reference/bed_shift.html">bed_shift()</a></code>: Shift the coordinates of an input set, bounded by a genome</li>
</ul><div id="sort" class="section level3">
<h3>Sort</h3>
<p><code>bed_sort</code> orders intervals based on a specification. <code>is_sorted</code> asks whether a tbl is already sorted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_sort.html">bed_sort</a></span>(x)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  5013  6013</span>
<span class="co">#&gt; 2   chr1  7141  8141</span>
<span class="co">#&gt; 3   chr1  7598  8598</span>
<span class="co">#&gt; 4   chr1 10256 11256</span>
<span class="co">#&gt; 5   chr1 12514 13514</span>
<span class="co">#&gt; 6   chr1 12912 13912</span>
<span class="co">#&gt; 7   chr1 13156 14156</span>
<span class="co">#&gt; 8   chr1 17374 18374</span>
<span class="co">#&gt; 9   chr1 26036 27036</span>
<span class="co">#&gt; 10  chr1 29659 30659</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(y)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
<div id="cluster" class="section level3">
<h3>Cluster</h3>
<p><code>bed_cluster</code> identifies clustered intervals based on a distance specification and assigns them a unique <code>.id</code>.</p>
<p><img src="README-cluster_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_cluster.html">bed_cluster</a></span>(x, <span class="dt">max_dist =</span> <span class="dv">1000</span>)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 4</span>
<span class="co">#&gt;    chrom     start       end   .id</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr9  22753037  22754037  7328</span>
<span class="co">#&gt; 2   chr3 131715169 131716169 42400</span>
<span class="co">#&gt; 3   chr4  89578351  89579351 28732</span>
<span class="co">#&gt; 4   chr5 136168027 136169027 43785</span>
<span class="co">#&gt; 5   chr3 150508207 150509207 48566</span>
<span class="co">#&gt; 6   chr1  50344463  50345463 16247</span>
<span class="co">#&gt; 7   chr7 107772322 107773322 34608</span>
<span class="co">#&gt; 8   chr8 115674629 115675629 37674</span>
<span class="co">#&gt; 9   chrY  28974494  28975494  9300</span>
<span class="co">#&gt; 10 chr20  24417930  24418930  7886</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="complement" class="section level3">
<h3>Complement</h3>
<p><code>bed_complement</code> identifies intervals in a genome that are not covered by an input.</p>
<p><img src="README-complement_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human)
<span class="co">#&gt; # A tibble: 724,411 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1     1 10295</span>
<span class="co">#&gt; 2   chr1 11295 14420</span>
<span class="co">#&gt; 3   chr1 15420 20055</span>
<span class="co">#&gt; 4   chr1 21055 29774</span>
<span class="co">#&gt; 5   chr1 30774 31917</span>
<span class="co">#&gt; 6   chr1 32917 34631</span>
<span class="co">#&gt; 7   chr1 35631 35937</span>
<span class="co">#&gt; 8   chr1 37769 41305</span>
<span class="co">#&gt; 9   chr1 42305 42676</span>
<span class="co">#&gt; 10  chr1 43760 44603</span>
<span class="co">#&gt; # ... with 724,401 more rows</span></code></pre></div>
</div>
<div id="merge" class="section level3">
<h3>Merge</h3>
<p><code>bed_merge</code> identifies overlapping intervals and reports new merged ones. <code>is_merged</code> asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.</p>
<p><img src="README-merge_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="fl">1e6</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> n)
<span class="kw"><a href="../reference/is_merged.html">is_merged</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

<span class="co"># add some signal</span>
x &lt;-<span class="st"> </span>x %&gt;%<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">signal =</span> <span class="kw">runif</span>(n))

<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x, <span class="dt">maxs =</span> <span class="kw">max</span>(signal))
<span class="co">#&gt; # A tibble: 723,485 &times; 4</span>
<span class="co">#&gt;    chrom start   end signal</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  1842  2842      .</span>
<span class="co">#&gt; 2   chr1  3671  4671      .</span>
<span class="co">#&gt; 3   chr1  5936  6936      .</span>
<span class="co">#&gt; 4   chr1 11170 12170      .</span>
<span class="co">#&gt; 5   chr1 20655 21655      .</span>
<span class="co">#&gt; 6   chr1 22516 24516      .</span>
<span class="co">#&gt; 7   chr1 26634 27634      .</span>
<span class="co">#&gt; 8   chr1 29170 30170      .</span>
<span class="co">#&gt; 9   chr1 35352 36393      .</span>
<span class="co">#&gt; 10  chr1 37783 38783      .</span>
<span class="co">#&gt; # ... with 723,475 more rows</span></code></pre></div>
</div>
<div id="flank" class="section level3">
<h3>Flank</h3>
<p><code>bed_flank</code> creates new intervals that flank &ndash; but do not contain &ndash; the input intervals.</p>
<p><img src="README-flank_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_flank.html">bed_flank</a></span>(x, genome_human, <span class="dt">both  =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    50   150</span>
<span class="co">#&gt; 2  chr1   125   225</span></code></pre></div>
</div>
<div id="slop" class="section level3">
<h3>Slop</h3>
<p><code>bed_slop</code> pads input intervals based on a specification</p>
<p><img src="README-slop_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_slop.html">bed_slop</a></span>(x, genome_human, <span class="dt">both =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    10   220</span>
<span class="co">#&gt; 2  chr1   125   335</span></code></pre></div>
</div>
<div id="shift" class="section level3">
<h3>Shift</h3>
<p><code>bed_shift</code> adjusts coordinates toward <code>start</code> or <code>end</code> by a defined <code>size</code>. Intervals created out of bounds are removed, or trimmed.</p>
<p><img src="README-shift_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_shift.html">bed_shift</a></span>(x, genome_human, <span class="dt">size =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1   125   150</span>
<span class="co">#&gt; 2  chr1   200   225</span></code></pre></div>
</div>
<div id="interval-spacing" class="section level3">
<h3>Interval spacing</h3>
<p>Interval spacing is easily computed using <code>dplyr</code>. Overlapping intervals must first be merged. Spacing for the first interval of each chromosome is undefined.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x)

x %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">.spacing =</span> start -<span class="st"> </span><span class="kw">lag</span>(end))
<span class="co">#&gt; Source: local data frame [723,348 x 4]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start   end .spacing</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   100  1100       NA</span>
<span class="co">#&gt; 2   chr1  3426  4426     2326</span>
<span class="co">#&gt; 3   chr1  5880  6880     1454</span>
<span class="co">#&gt; 4   chr1  7069  8069      189</span>
<span class="co">#&gt; 5   chr1  8230  9280      161</span>
<span class="co">#&gt; 6   chr1 13183 14183     3903</span>
<span class="co">#&gt; 7   chr1 18418 19418     4235</span>
<span class="co">#&gt; 8   chr1 23828 24828     4410</span>
<span class="co">#&gt; 9   chr1 25372 26372      544</span>
<span class="co">#&gt; 10  chr1 30283 32145     3911</span>
<span class="co">#&gt; # ... with 723,338 more rows</span></code></pre></div>
</div>
</div>
<div id="multiple-set-operations" class="section level2 tabset">
<h2>Multiple set operations</h2>
<p>These methods compare two sets of intervals:</p>
<ul><li><code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>: find overlapping intervals</li>
<li><code><a href="../reference/bed_map.html">bed_map()</a></code>: apply a function to selected columns for overlapping intervals</li>
<li><code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>: Remove intervals based on overlaps between two files</li>
<li><code><a href="../reference/bed_window.html">bed_window()</a></code>: Find overlapping intervals within a window</li>
<li><code><a href="../reference/bed_closest.html">bed_closest()</a></code>: find the closest intervals independent of overlaps</li>
</ul><div id="intersection" class="section level3">
<h3>Intersection</h3>
<p><code>bed_intersect</code> is implemented using an <a href="https://github.com/ekg/intervaltree/">interval tree</a> in <code>Rcpp</code>. Column names in the result have <code>.x</code> and <code>.y</code> suffixes, and an <code>.overlap</code> column contains the size of the intersection (values of <code>0</code> indicate book-ended, or touching intervals). See the <a href="#benchmarks">benchmarks</a> section for timing. Though <code>bed_intersect</code> is pretty fast already, we intend to further improve upon this by parallization with <a href="https://rcppcore.github.io/RcppParallel/"><code>RcppParallel</code></a>.</p>
<p><img src="README-intersect_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># intersect two sets of 1e6 intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 645,764 &times; 6</span>
<span class="co">#&gt;    chrom start.x end.x start.y end.y .overlap</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt; &lt;int&gt;   &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1    8225  9225    8453  9453      772</span>
<span class="co">#&gt; 2   chr1   15610 16610   15202 16202      592</span>
<span class="co">#&gt; 3   chr1   15974 16974   15202 16202      228</span>
<span class="co">#&gt; 4   chr1   19249 20249   19568 20568      681</span>
<span class="co">#&gt; 5   chr1   23822 24822   24171 25171      651</span>
<span class="co">#&gt; 6   chr1   24374 25374   24171 25171      797</span>
<span class="co">#&gt; 7   chr1   43117 44117   42857 43857      740</span>
<span class="co">#&gt; 8   chr1   45206 46206   45160 46160      954</span>
<span class="co">#&gt; 9   chr1   51639 52639   52550 53550       89</span>
<span class="co">#&gt; 10  chr1   52820 53820   52550 53550      730</span>
<span class="co">#&gt; # ... with 645,754 more rows</span>

<span class="co"># A records with no overlaps (i.e., `-v`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y, <span class="dt">invert =</span> <span class="ot">TRUE</span>)
<span class="co">#&gt; # A tibble: 524,364 &times; 3</span>
<span class="co">#&gt;    chrom    start      end</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chrY 59362122 59363122</span>
<span class="co">#&gt; 2   chrY 59359267 59360267</span>
<span class="co">#&gt; 3   chrY 59353251 59354251</span>
<span class="co">#&gt; 4   chrY 59352052 59353052</span>
<span class="co">#&gt; 5   chrY 59339447 59340447</span>
<span class="co">#&gt; 6   chrY 59339140 59340140</span>
<span class="co">#&gt; 7   chrY 59338220 59339220</span>
<span class="co">#&gt; 8   chrY 59338170 59339170</span>
<span class="co">#&gt; 9   chrY 59331219 59332219</span>
<span class="co">#&gt; 10  chrY 59320202 59321202</span>
<span class="co">#&gt; # ... with 524,354 more rows</span></code></pre></div>
<p>One can achieve behaviour similar to BEDtools by combining <code>bed_intersect</code> with <code>dplyr</code> tools.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># `x` records with overlaps (i.e., `-wa`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x)
<span class="co">#&gt; # A tibble: 645,764 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  8225  9225</span>
<span class="co">#&gt; 2   chr1 15610 16610</span>
<span class="co">#&gt; 3   chr1 15974 16974</span>
<span class="co">#&gt; 4   chr1 19249 20249</span>
<span class="co">#&gt; 5   chr1 23822 24822</span>
<span class="co">#&gt; 6   chr1 24374 25374</span>
<span class="co">#&gt; 7   chr1 43117 44117</span>
<span class="co">#&gt; 8   chr1 45206 46206</span>
<span class="co">#&gt; 9   chr1 51639 52639</span>
<span class="co">#&gt; 10  chr1 52820 53820</span>
<span class="co">#&gt; # ... with 645,754 more rows</span>

<span class="co"># `y` records with overlaps (i.e., `-wb`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.y, <span class="dt">end =</span> end.y)
<span class="co">#&gt; # A tibble: 645,764 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  8453  9453</span>
<span class="co">#&gt; 2   chr1 15202 16202</span>
<span class="co">#&gt; 3   chr1 15202 16202</span>
<span class="co">#&gt; 4   chr1 19568 20568</span>
<span class="co">#&gt; 5   chr1 24171 25171</span>
<span class="co">#&gt; 6   chr1 24171 25171</span>
<span class="co">#&gt; 7   chr1 42857 43857</span>
<span class="co">#&gt; 8   chr1 45160 46160</span>
<span class="co">#&gt; 9   chr1 52550 53550</span>
<span class="co">#&gt; 10  chr1 52550 53550</span>
<span class="co">#&gt; # ... with 645,754 more rows</span>

<span class="co"># Unique records in `x` (i.e., `-u`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x) %&gt;%<span class="st"> </span><span class="kw">unique</span>()
<span class="co">#&gt; # A tibble: 475,548 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  8225  9225</span>
<span class="co">#&gt; 2   chr1 15610 16610</span>
<span class="co">#&gt; 3   chr1 15974 16974</span>
<span class="co">#&gt; 4   chr1 19249 20249</span>
<span class="co">#&gt; 5   chr1 23822 24822</span>
<span class="co">#&gt; 6   chr1 24374 25374</span>
<span class="co">#&gt; 7   chr1 43117 44117</span>
<span class="co">#&gt; 8   chr1 45206 46206</span>
<span class="co">#&gt; 9   chr1 51639 52639</span>
<span class="co">#&gt; 10  chr1 52820 53820</span>
<span class="co">#&gt; # ... with 475,538 more rows</span>

<span class="co"># Count `y` intervals that intersect each `x` interval</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">group_by</span>(chrom, start.x, end.x) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">count =</span> <span class="kw">n</span>())
<span class="co">#&gt; Source: local data frame [475,548 x 4]</span>
<span class="co">#&gt; Groups: chrom, start.x [?]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start.x end.x count</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1    8225  9225     1</span>
<span class="co">#&gt; 2   chr1   15610 16610     1</span>
<span class="co">#&gt; 3   chr1   15974 16974     1</span>
<span class="co">#&gt; 4   chr1   19249 20249     1</span>
<span class="co">#&gt; 5   chr1   23822 24822     1</span>
<span class="co">#&gt; 6   chr1   24374 25374     1</span>
<span class="co">#&gt; 7   chr1   43117 44117     1</span>
<span class="co">#&gt; 8   chr1   45206 46206     1</span>
<span class="co">#&gt; 9   chr1   51639 52639     1</span>
<span class="co">#&gt; 10  chr1   52820 53820     2</span>
<span class="co">#&gt; # ... with 475,538 more rows</span></code></pre></div>
</div>
<div id="map" class="section level3">
<h3>Map</h3>
<p><code>bed_map</code> maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs.</p>
<p><img src="README-map_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bedfile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)
bgfile  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)

x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(bgfile)

<span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">means =</span> <span class="kw">mean</span>(value), <span class="dt">sds =</span> <span class="kw">sd</span>(value))
<span class="co">#&gt; # A tibble: 730 &times; 8</span>
<span class="co">#&gt;    chrom    start      end    means      sds     name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr22 16150259 16193004 7.914286 7.570731     &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 2  chr22 16162065 16172265 1.000000       NA     &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 3  chr22 16179617 16181004       NA       NA BC017398     1      -</span>
<span class="co">#&gt; 4  chr22 16239287 16239327       NA       NA DQ590589     1      +</span>
<span class="co">#&gt; 5  chr22 16240245 16240277       NA       NA DQ573684     1      -</span>
<span class="co">#&gt; 6  chr22 16240300 16240340       NA       NA DQ595048     1      -</span>
<span class="co">#&gt; 7  chr22 16241085 16241125       NA       NA DQ590589     1      +</span>
<span class="co">#&gt; 8  chr22 16242753 16242785       NA       NA DQ573684     1      -</span>
<span class="co">#&gt; 9  chr22 16243382 16243414       NA       NA DQ573684     1      -</span>
<span class="co">#&gt; 10 chr22 16243447 16243487       NA       NA DQ587539     1      -</span>
<span class="co">#&gt; # ... with 720 more rows</span></code></pre></div>
</div>
<div id="subtract" class="section level3">
<h3>Subtract</h3>
<p><code>bed_substract()</code> removes <code>x</code> intervals that intersect with <code>y</code>.</p>
<p><img src="README-subtract_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y)
<span class="co">#&gt; # A tibble: 957,540 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   452  1452</span>
<span class="co">#&gt; 2   chr1  1223  2223</span>
<span class="co">#&gt; 3   chr1  6806  7567</span>
<span class="co">#&gt; 4   chr1  7168  7567</span>
<span class="co">#&gt; 5   chr1  8567  8919</span>
<span class="co">#&gt; 6   chr1 13070 13738</span>
<span class="co">#&gt; 7   chr1 15771 16771</span>
<span class="co">#&gt; 8   chr1 20913 21304</span>
<span class="co">#&gt; 9   chr1 22645 23645</span>
<span class="co">#&gt; 10  chr1 29131 29476</span>
<span class="co">#&gt; # ... with 957,530 more rows</span></code></pre></div>
</div>
<div id="window" class="section level3">
<h3>Window</h3>
<p><code><a href="../reference/bed_window.html">bed_window()</a></code> identifies <code>y</code> intervals that intersect an expanded window of <code>x</code> intervals.</p>
<p><img src="README-window_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="co"># a few intersections</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span>

<span class="co"># can be expanded by casting a wider net</span>
<span class="kw"><a href="../reference/bed_window.html">bed_window</a></span>(x, y, genome, <span class="dt">both =</span> <span class="fl">1e6</span>)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span></code></pre></div>
</div>
<div id="closest" class="section level3">
<h3>Closest</h3>
<p><code><a href="../reference/bed_closest.html">bed_closest()</a></code> identifies <code>y</code> intervals that are closest to <code>x</code>.</p>
<p><img src="README-closest_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y)
<span class="co">#&gt; # A tibble: 98 &times; 7</span>
<span class="co">#&gt;    chrom   start.x     end.x   start.y     end.y .overlap .distance</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;    &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  14714752  14715752  14056671  14057671        0   -657081</span>
<span class="co">#&gt; 2   chr1  43935036  43936036  46067807  46068807        0   2131771</span>
<span class="co">#&gt; 3   chr1  54517016  54518016  46067807  46068807        0  -8448209</span>
<span class="co">#&gt; 4   chr1  81312928  81313928  81793362  81794362        0    479434</span>
<span class="co">#&gt; 5   chr1  82009400  82010400  81793362  81794362        0   -215038</span>
<span class="co">#&gt; 6   chr1 119616128 119617128 126469777 126470777        0   6852649</span>
<span class="co">#&gt; 7   chr1 175997140 175998140 173684076 173685076        0  -2312064</span>
<span class="co">#&gt; 8  chr10   1594124   1595124  16933129  16934129        0  15338005</span>
<span class="co">#&gt; 9  chr10   3459689   3460689  16933129  16934129        0  13472440</span>
<span class="co">#&gt; 10 chr11   8563700   8564700   2953228   2954228        0  -5609472</span>
<span class="co">#&gt; # ... with 88 more rows</span></code></pre></div>
</div>
</div>
<div id="randomzing-intervals" class="section level2 tabset">
<h2>Randomzing intervals</h2>
<p><code>valr</code> provides methods for creating new random intervals or permutations of existing intervals:</p>
<ul><li><code>bed_random</code> generates random intervals from an input <code>genome</code>.</li>
<li><code>bed_shuffle</code> shuffles coordinates given a set of input intervals.</li>
<li>Random sampling of input intervals is done with <code>dplyr</code>.</li>
</ul><div id="random" class="section level3">
<h3>Random</h3>
<p><code>bed_random</code> generates random intervals from an input genome. The numbers of intervals from each <code>chrom</code> are proporitional to each chrom <code>size</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="fl">1e6</span>, <span class="dt">length =</span> <span class="fl">1e3</span>)
x
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr4   5996090   5997090</span>
<span class="co">#&gt; 2   chr1 114018912 114019912</span>
<span class="co">#&gt; 3  chr12  98743840  98744840</span>
<span class="co">#&gt; 4   chr8  72057836  72058836</span>
<span class="co">#&gt; 5   chr7 154595396 154596396</span>
<span class="co">#&gt; 6  chr14  42606402  42607402</span>
<span class="co">#&gt; 7   chr1  12596556  12597556</span>
<span class="co">#&gt; 8  chr19  56249872  56250872</span>
<span class="co">#&gt; 9  chr12  20668622  20669622</span>
<span class="co">#&gt; 10 chr10 134679278 134680278</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="co"># numbers of sampled intervals are proportional to chrom size</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">n =</span> <span class="kw">n</span>()) %&gt;%<span class="st"> </span><span class="kw">arrange</span>(<span class="kw">desc</span>(n))
<span class="co">#&gt; # A tibble: 25 &times; 2</span>
<span class="co">#&gt;    chrom     n</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 81022</span>
<span class="co">#&gt; 2   chr2 78662</span>
<span class="co">#&gt; 3   chr3 64382</span>
<span class="co">#&gt; 4   chr4 61520</span>
<span class="co">#&gt; 5   chr5 58543</span>
<span class="co">#&gt; 6   chr6 55343</span>
<span class="co">#&gt; 7   chr7 51082</span>
<span class="co">#&gt; 8   chrX 50223</span>
<span class="co">#&gt; 9   chr8 47165</span>
<span class="co">#&gt; 10  chr9 45638</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="sample" class="section level3">
<h3>Sample</h3>
<p>Sampling can be done using <code>dplyr</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="co"># sample by number</span>
<span class="kw">sample_n</span>(x, <span class="fl">1e3</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 1,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr11  50058172  50059172</span>
<span class="co">#&gt; 2   chr8  35221753  35222753</span>
<span class="co">#&gt; 3   chrX   4647643   4648643</span>
<span class="co">#&gt; 4   chr6 150346558 150347558</span>
<span class="co">#&gt; 5   chr7 128581398 128582398</span>
<span class="co">#&gt; 6   chr2 213233434 213234434</span>
<span class="co">#&gt; 7  chr12  91415503  91416503</span>
<span class="co">#&gt; 8   chr9  94578657  94579657</span>
<span class="co">#&gt; 9   chr3 117386899 117387899</span>
<span class="co">#&gt; 10  chr5  18809495  18810495</span>
<span class="co">#&gt; # ... with 990 more rows</span>

<span class="co"># or fraction</span>
<span class="kw">sample_frac</span>(x, <span class="fl">0.1</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 100,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr16  66037123  66038123</span>
<span class="co">#&gt; 2  chr16   9312032   9313032</span>
<span class="co">#&gt; 3  chr12  76686706  76687706</span>
<span class="co">#&gt; 4   chr9 115777346 115778346</span>
<span class="co">#&gt; 5   chrY  24243985  24244985</span>
<span class="co">#&gt; 6   chrX  72602071  72603071</span>
<span class="co">#&gt; 7   chr8  47000806  47001806</span>
<span class="co">#&gt; 8   chr3 177903558 177904558</span>
<span class="co">#&gt; 9   chr8  45695450  45696450</span>
<span class="co">#&gt; 10 chr13  67478195  67479195</span>
<span class="co">#&gt; # ... with 99,990 more rows</span>

<span class="co"># or sample intervals within groups</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">sample_n</span>(<span class="dv">1</span>)
<span class="co">#&gt; Source: local data frame [25 x 3]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 124561857 124562857</span>
<span class="co">#&gt; 2  chr10  91516195  91517195</span>
<span class="co">#&gt; 3  chr11  80843114  80844114</span>
<span class="co">#&gt; 4  chr12  50609244  50610244</span>
<span class="co">#&gt; 5  chr13  48655053  48656053</span>
<span class="co">#&gt; 6  chr14  58019341  58020341</span>
<span class="co">#&gt; 7  chr15 101083681 101084681</span>
<span class="co">#&gt; 8  chr16  25023030  25024030</span>
<span class="co">#&gt; 9  chr17  29565844  29566844</span>
<span class="co">#&gt; 10 chr18  34162912  34163912</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="shuffle" class="section level3">
<h3>Shuffle</h3>
<p><code>bed_shuffle</code> shuffles input intervals. Interval sizes are equal in the input and output.</p>
<pre><code>#&gt; Warning in bind_rows_(x, .id): binding factor and character vector,
#&gt; coercing into character vector</code></pre>
<p><img src="README-shuffle_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human)
y
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;    chrom    start      end</span>
<span class="co">#&gt;   &lt;fctr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1  chr20 27653018 27653038</span>
<span class="co">#&gt; 2  chr10 50132221 50132231</span>

<span class="kw">any</span>(x$start ==<span class="st"> </span>y$start)
<span class="co">#&gt; [1] FALSE</span>

<span class="kw">all</span>(x$end -<span class="st"> </span>x$start ==<span class="st"> </span>y$end -<span class="st"> </span>y$start)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
</div>
<div id="other-topics" class="section level2">
<h2>Other topics</h2>
<div id="grouping-related-intervals" class="section level3">
<h3>Grouping related intervals</h3>
<p>The <code>group_by</code> operation from <code>dplyr</code> serves a similar purpose to BEDtools <code>groupby</code>, and can be combined with <code>dplyr::summarize</code> to calculate summary statistics from groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">overlap.total =</span> <span class="kw">sum</span>(.overlap))
<span class="co">#&gt; # A tibble: 25 &times; 2</span>
<span class="co">#&gt;    chrom overlap.total</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1      26061005</span>
<span class="co">#&gt; 2  chr10      14168388</span>
<span class="co">#&gt; 3  chr11      14063707</span>
<span class="co">#&gt; 4  chr12      13961668</span>
<span class="co">#&gt; 5  chr13      12031014</span>
<span class="co">#&gt; 6  chr14      11252334</span>
<span class="co">#&gt; 7  chr15      10545980</span>
<span class="co">#&gt; 8  chr16       9281860</span>
<span class="co">#&gt; 9  chr17       8416537</span>
<span class="co">#&gt; 10 chr18       8077764</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
</div>
</div>
<div id="interval-statistics" class="section level1 tabset">
<h1>Interval Statistics</h1>
<p><code>valr</code> provides several methods to assess statistical properties of interval sets including:</p>
<ul><li><code><a href="../reference/bed_fisher.html">bed_fisher()</a></code>: measure overlap significance of two sets of intervals.</li>
<li><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>: quantify absolute distance between query intervals and closest reference intervals</li>
<li><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code>: quantify relative distances between query intervals and closest reference intervals</li>
<li><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>: quantify extent of overlap between two sets of intervals</li>
</ul><p>Several of these methods were described in the <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">Genometricorr</a> software package.</p>
<div id="fishers-test" class="section level2">
<h2>Fisher&rsquo;s test</h2>
<p>The Fisher&rsquo;s test assesses whether two sets of intervals are drawn from the same background genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_fisher.html">bed_fisher</a></span>(x, y, genome_human)
<span class="co">#&gt;   estimate p.value conf.low conf.high                             method</span>
<span class="co">#&gt; 1 2.353602       0 2.339484  2.367925 Fisher's Exact Test for Count Data</span>
<span class="co">#&gt;   alternative</span>
<span class="co">#&gt; 1   two.sided</span></code></pre></div>
</div>
<div id="absolute-distance" class="section level2">
<h2>Absolute distance</h2>
<p><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code> computes the absolute distance between the midpoint of query intervals and the closest midpoints of a set of reference intervals. Absolute distances are scaled by the inter-reference gap for the chromosome as follows. For <code>Q</code> total query points and <code>R</code> reference points on a chromosome, scale the distance for each query point <code>i</code> to the closest reference point by the inter-reference gap for each chromosome.</p>
<p><span class="math display">\[
d_i(x,y) = min_{\substack{k}}(|q_i - r_k|)\frac{R}{Length\ of\ chromsome}
\]</span></p>
<p>By default both absolute and scaled distances are reported as <code>absdist</code> and <code>scaled_absdist</code> respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_absdist.html">bed_absdist</a></span>(x, y, genome_human)
<span class="co">#&gt; # A tibble: 1,000,000 &times; 5</span>
<span class="co">#&gt;    chrom start   end absdist scaled_absdist</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;   &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1   chr1  2016  3016    2412     0.77782814</span>
<span class="co">#&gt; 2   chr1 11083 12083     149     0.04804991</span>
<span class="co">#&gt; 3   chr1 12535 13535     521     0.16801346</span>
<span class="co">#&gt; 4   chr1 15974 16974     744     0.23992709</span>
<span class="co">#&gt; 5   chr1 16882 17882     271     0.08739280</span>
<span class="co">#&gt; 6   chr1 19906 20906    2612     0.84232467</span>
<span class="co">#&gt; 7   chr1 22151 23151     367     0.11835113</span>
<span class="co">#&gt; 8   chr1 23141 24141     188     0.06062674</span>
<span class="co">#&gt; 9   chr1 23678 24678     349     0.11254644</span>
<span class="co">#&gt; 10  chr1 25217 26217     337     0.10867665</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="relative-distance" class="section level2">
<h2>Relative distance</h2>
<p><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> computes the relative distance between a query interval and the two closest reference intervals. The relative distance can range between <code>[0,0.5]</code>. If the overall distribution of distances is biased toward zero, then the query and reference intervals tend to be closer in space (<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">17</a>). By default a summary with the frequency of each relative distance is reported.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_reldist.html">bed_reldist</a></span>(x, y)
<span class="co">#&gt; # A tibble: 51 &times; 4</span>
<span class="co">#&gt;    reldist counts  total       freq</span>
<span class="co">#&gt;      &lt;dbl&gt;  &lt;int&gt;  &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1     0.00  20169 999948 0.02017005</span>
<span class="co">#&gt; 2     0.01  19726 999948 0.01972703</span>
<span class="co">#&gt; 3     0.02  19952 999948 0.01995304</span>
<span class="co">#&gt; 4     0.03  20326 999948 0.02032706</span>
<span class="co">#&gt; 5     0.04  19580 999948 0.01958102</span>
<span class="co">#&gt; 6     0.05  19994 999948 0.01999504</span>
<span class="co">#&gt; 7     0.06  19968 999948 0.01996904</span>
<span class="co">#&gt; 8     0.07  20439 999948 0.02044006</span>
<span class="co">#&gt; 9     0.08  20004 999948 0.02000504</span>
<span class="co">#&gt; 10    0.09  19829 999948 0.01983003</span>
<span class="co">#&gt; # ... with 41 more rows</span></code></pre></div>
</div>
<div id="jaccard-similarity" class="section level2">
<h2>Jaccard similarity</h2>
<p><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code> quantifies the extent of overlap between to sets of intervals. The Jaccard statistic takes values of <code>[0,1]</code> and is measured as:</p>
<p><span class="math display">\[
J(x,y) = \frac{\mid x \bigcap y \mid}
              {\mid x \bigcup y \mid} 
       = 
         \frac{\mid x \bigcap y \mid}
              {\mid x \mid + \mid y \mid - \mid x \bigcap y \mid}
\]</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_jaccard.html">bed_jaccard</a></span>(x, y)
<span class="co">#&gt; # A tibble: 1 &times; 4</span>
<span class="co">#&gt;       len_i      len_u   jaccard      n</span>
<span class="co">#&gt;       &lt;int&gt;      &lt;int&gt;     &lt;dbl&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1 323210177 2000000000 0.1927553 646084</span></code></pre></div>
</div>
</div>
<div id="benchmarks" class="section level1">
<h1>Benchmarks</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># two sets of 1e6 random 1 kb intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw">library</span>(microbenchmark)
<span class="kw">microbenchmark</span>(
  <span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human),
  <span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x),
  <span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human),
  <span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human),
  <span class="dt">times =</span> <span class="dv">1</span>,
  <span class="dt">unit =</span> <span class="st">'s'</span>
)
<span class="co">#&gt; Unit: seconds</span>
<span class="co">#&gt;                             expr      min       lq     mean   median</span>
<span class="co">#&gt;         bed_random(genome_human) 0.112214 0.112214 0.112214 0.112214</span>
<span class="co">#&gt;                bed_closest(x, y) 5.114431 5.114431 5.114431 5.114431</span>
<span class="co">#&gt;              bed_intersect(x, y) 1.722588 1.722588 1.722588 1.722588</span>
<span class="co">#&gt;                     bed_merge(x) 2.207061 2.207061 2.207061 2.207061</span>
<span class="co">#&gt;               bed_subtract(x, y) 3.722285 3.722285 3.722285 3.722285</span>
<span class="co">#&gt;  bed_complement(x, genome_human) 3.012574 3.012574 3.012574 3.012574</span>
<span class="co">#&gt;     bed_shuffle(x, genome_human) 1.128805 1.128805 1.128805 1.128805</span>
<span class="co">#&gt;        uq      max neval</span>
<span class="co">#&gt;  0.112214 0.112214     1</span>
<span class="co">#&gt;  5.114431 5.114431     1</span>
<span class="co">#&gt;  1.722588 1.722588     1</span>
<span class="co">#&gt;  2.207061 2.207061     1</span>
<span class="co">#&gt;  3.722285 3.722285     1</span>
<span class="co">#&gt;  3.012574 3.012574     1</span>
<span class="co">#&gt;  1.128805 1.128805     1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#overview">Overview</a><ul class="nav nav-pills nav-stacked"><li><a href="#installation">Installation</a></li>
      <li><a href="#comparison-to-other-tools">Comparison to other tools</a></li>
      <li><a href="#non-standard-evaluation">Non-standard evaluation</a></li>
      <li><a href="#getting-started">Getting started</a></li>
      </ul></li>
      <li><a href="#reading-data">Reading data</a><ul class="nav nav-pills nav-stacked"><li><a href="#bed-files">BED files</a></li>
      <li><a href="#genome-files">Genome files</a></li>
      <li><a href="#bedgraph-files">bedGraph files</a></li>
      <li><a href="#vcf-files">VCF files</a></li>
      </ul></li>
      <li><a href="#interval-manipulations">Interval manipulations</a><ul class="nav nav-pills nav-stacked"><li><a href="#single-set-operations">Single set operations</a></li>
      <li><a href="#multiple-set-operations">Multiple set operations</a></li>
      <li><a href="#randomzing-intervals">Randomzing intervals</a></li>
      <li><a href="#other-topics">Other topics</a></li>
      </ul></li>
      <li><a href="#interval-statistics">Interval Statistics</a><ul class="nav nav-pills nav-stacked"><li><a href="#fishers-test">Fisher&rsquo;s test</a></li>
      <li><a href="#absolute-distance">Absolute distance</a></li>
      <li><a href="#relative-distance">Relative distance</a></li>
      <li><a href="#jaccard-similarity">Jaccard similarity</a></li>
      </ul></li>
      <li><a href="#benchmarks">Benchmarks</a></li>
      </ul></div>
      </div>

</div>


      <footer><p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer></div>

  </body></html>
