<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>valr functions &bull; valr</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container">
      <header><div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/valr-examples.html">Examples</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>valr functions</h1>
                        <h4 class="author">Jay Hesselberth <a href="mailto:jay.hesselberth@gmail.com">jay.hesselberth@gmail.com</a></h4>
            
            <h4 class="date">2016-10-18</h4>
          </div>

    
    
<div id="reading-data" class="section level1 tabset">
<h1>Reading data</h1>
<p><code>valr</code> has several methods to read interval data. These methods:</p>
<ul><li>Take local files or URLs as input.</li>
<li>Return a <code>data.frame</code> in <code>tibble::tbl_df</code> format.</li>
<li>Assign consistent <code>chrom</code>, <code>start</code> and <code>end</code> column names.</li>
<li>Use <a href="https://github.com/hadley/readr"><code>readr</code></a> for speed.</li>
<li>Coerce column types.</li>
</ul><p>The methods include:</p>
<ul><li><code><a href="../reference/read_bed.html">read_bed()</a></code>: read a BED3+ file</li>
<li><code><a href="../reference/read_bed.html">read_bed12()</a></code>: read a BED12 file</li>
<li><code><a href="../reference/read_bed.html">read_bedgraph()</a></code>: read a bedGraph file</li>
<li><code><a href="../reference/read_genome.html">read_genome()</a></code>: read a UCSC &ldquo;chrom size&rdquo; file</li>
<li><code><a href="../reference/read_vcf.html">read_vcf()</a></code>: read the Variant Call Format</li>
<li><code><a href="../reference/read_bed.html">read_narrowpeak()</a></code>: read narrowPeak files</li>
<li><code><a href="../reference/read_bed.html">read_broadpeak()</a></code>: read broadPeak files</li>
</ul><div id="bed-files" class="section level2">
<h2>BED files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'3fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 3</span>
<span class="co">#&gt;    chrom  start    end</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409</span>
<span class="co">#&gt; 2   chr1  14361  19759</span>
<span class="co">#&gt; 3   chr1  14406  29370</span>
<span class="co">#&gt; 4   chr1  34610  36081</span>
<span class="co">#&gt; 5   chr1  69090  70008</span>
<span class="co">#&gt; 6   chr1 134772 140566</span>
<span class="co">#&gt; 7   chr1 321083 321115</span>
<span class="co">#&gt; 8   chr1 321145 321207</span>
<span class="co">#&gt; 9   chr1 322036 326938</span>
<span class="co">#&gt; 10  chr1 327545 328439</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="dt">n_fields =</span> <span class="dv">6</span>, <span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'6fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409      DDX11L1     3      +</span>
<span class="co">#&gt; 2   chr1  14361  19759       WASH7P    10      -</span>
<span class="co">#&gt; 3   chr1  14406  29370       WASH7P     7      -</span>
<span class="co">#&gt; 4   chr1  34610  36081      FAM138F     3      -</span>
<span class="co">#&gt; 5   chr1  69090  70008        OR4F5     1      +</span>
<span class="co">#&gt; 6   chr1 134772 140566    LOC729737     3      -</span>
<span class="co">#&gt; 7   chr1 321083 321115     DQ597235     1      +</span>
<span class="co">#&gt; 8   chr1 321145 321207     DQ599768     1      +</span>
<span class="co">#&gt; 9   chr1 322036 326938 LOC100133331     3      +</span>
<span class="co">#&gt; 10  chr1 327545 328439    LOC388312     1      +</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed12</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'12fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 3 &times; 12</span>
<span class="co">#&gt;   chrom   start     end      name score strand cds_start cds_end item_rgb</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;   &lt;int&gt;    &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1 4797973 4836816  testgene     1      +   4797973 4836816        .</span>
<span class="co">#&gt; 2 chr10 4848118 4880877 diffchrom     1      +   4848118 4880877        .</span>
<span class="co">#&gt; 3 chr20 5073253 5152630 negstrand     1      -   5073253 5152630        .</span>
<span class="co">#&gt; # ... with 3 more variables: exon_count &lt;int&gt;, exon_sizes &lt;chr&gt;,</span>
<span class="co">#&gt; #   exon_starts &lt;chr&gt;</span></code></pre></div>
</div>
<div id="genome-files" class="section level2">
<h2>Genome files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome_human &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>))</code></pre></div>
</div>
<div id="bedgraph-files" class="section level2">
<h2>bedGraph files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.bg.gz'</span>))
<span class="co">#&gt; # A tibble: 4 &times; 4</span>
<span class="co">#&gt;   chrom    start      end value</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr19 49302000 49302300 -1.00</span>
<span class="co">#&gt; 2 chr19 49302300 49302600 -0.75</span>
<span class="co">#&gt; 3 chr19 49302600 49302900 -0.50</span>
<span class="co">#&gt; 4 chr19 49302900 49303200 -0.25</span></code></pre></div>
</div>
<div id="vcf-files" class="section level2">
<h2>VCF files</h2>
<p><code><a href="../reference/read_vcf.html">read_vcf()</a></code> reads VCF files and assigns <code>chrom</code>, <code>start</code> and <code>end</code> columns to be used to downstream interval comparisons. Note the interval size is calculated as the length of the <code>REF</code> field in the original file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_vcf.html">read_vcf</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.vcf.gz'</span>))
<span class="co">#&gt; # A tibble: 11 &times; 18</span>
<span class="co">#&gt;    CHROM   POS      ID   REF   ALT  QUAL FILTER</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1      1    10    1:10     A     T   100   PASS</span>
<span class="co">#&gt; 2      1    20    1:20     G     C   100   PASS</span>
<span class="co">#&gt; 3      1    30    1:30     C     A   100   PASS</span>
<span class="co">#&gt; 4      1    40    1:40     A     C   100   PASS</span>
<span class="co">#&gt; 5      1 10000 1:10000     G     C   100   PASS</span>
<span class="co">#&gt; 6      1 20000 1:20000     T     A   100   PASS</span>
<span class="co">#&gt; 7      4  5000  4:5000     A     T   100   PASS</span>
<span class="co">#&gt; 8      4  6000  4:6000     C     T   100   PASS</span>
<span class="co">#&gt; 9      X   800   X:800     A     C   100   PASS</span>
<span class="co">#&gt; 10     X   900   X:900     A     T   100   PASS</span>
<span class="co">#&gt; 11     X  1000  X:1000     T     G   100   PASS</span>
<span class="co">#&gt; # ... with 11 more variables: INFO &lt;chr&gt;, FORMAT &lt;chr&gt;, X1 &lt;chr&gt;,</span>
<span class="co">#&gt; #   X2 &lt;chr&gt;, X3 &lt;chr&gt;, X4 &lt;chr&gt;, X5 &lt;chr&gt;, X6 &lt;chr&gt;, chrom &lt;chr&gt;,</span>
<span class="co">#&gt; #   start &lt;int&gt;, end &lt;int&gt;</span></code></pre></div>
</div>
</div>
<div id="interval-manipulations" class="section level1">
<h1>Interval manipulations</h1>
<p><code>valr</code> implements several methods for manipulating sets of intervals. Some methods operate on a single set of intervals, while others compare two sets of intervals.</p>
<p>Many methods the same name as the corresponding <code>BEDtool</code>, and some commonly used <code>BEDtools</code> are implemented as <code>dplyr</code> pipes (e.g., see the <a href="#groupby"><code>group_by</code></a> section).</p>
<p>All methods accept one or more sets of <code>x</code> and <code>y</code> intervals, which must either be created using the <a href="#reading%20data"><code>read</code> methods</a>, or have <code>chrom</code>, <code>start</code> and <code>end</code> columns.</p>
<div id="single-set-operations" class="section level2 tabset">
<h2>Single set operations</h2>
<p>These methods operate on a single set of intervals:</p>
<ul><li><code><a href="../reference/bed_sort.html">bed_sort()</a></code>: order intervals</li>
<li><code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>: Cluster (but don&rsquo;t merge) overlapping/nearby intervals.</li>
<li><code><a href="../reference/bed_complement.html">bed_complement()</a></code>: extract intervals <em>not</em> represented by an interval file.</li>
<li><code><a href="../reference/bed_merge.html">bed_merge()</a></code>: combine overlapping and nearby intervals into a single interval.</li>
<li><code><a href="../reference/bed_flank.html">bed_flank()</a></code>: Generate new flanking intervals</li>
<li><code><a href="../reference/bed_slop.html">bed_slop()</a></code>: Expand the size of input intervals</li>
<li><code><a href="../reference/bed_shift.html">bed_shift()</a></code>: Shift the coordinates of an input set, bounded by a genome</li>
</ul><div id="sort" class="section level3">
<h3>Sort</h3>
<p><code>bed_sort</code> orders intervals based on a specification. <code>is_sorted</code> asks whether a tbl is already sorted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_sort.html">bed_sort</a></span>(x)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1    66  1066</span>
<span class="co">#&gt; 2   chr1  4656  5656</span>
<span class="co">#&gt; 3   chr1  5417  6417</span>
<span class="co">#&gt; 4   chr1  6634  7634</span>
<span class="co">#&gt; 5   chr1  7413  8413</span>
<span class="co">#&gt; 6   chr1  8513  9513</span>
<span class="co">#&gt; 7   chr1 10740 11740</span>
<span class="co">#&gt; 8   chr1 14983 15983</span>
<span class="co">#&gt; 9   chr1 24041 25041</span>
<span class="co">#&gt; 10  chr1 24502 25502</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(y)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
<div id="cluster" class="section level3">
<h3>Cluster</h3>
<p><code>bed_cluster</code> identifies clustered intervals based on a distance specification and assigns them a unique <code>.id</code>.</p>
<p><img src="img/functions-cluster_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_cluster.html">bed_cluster</a></span>(x, <span class="dt">max_dist =</span> <span class="dv">1000</span>)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 4</span>
<span class="co">#&gt;    chrom     start       end   .id</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chrY  40255269  40256269 12956</span>
<span class="co">#&gt; 2   chrX 114616800 114617800 36979</span>
<span class="co">#&gt; 3  chr11  90606200  90607200 29392</span>
<span class="co">#&gt; 4  chr14  13847261  13848261  4415</span>
<span class="co">#&gt; 5  chr18  16493745  16494745  5372</span>
<span class="co">#&gt; 6  chr10 111986816 111987816 36012</span>
<span class="co">#&gt; 7  chr22  27321558  27322558  8822</span>
<span class="co">#&gt; 8   chr3  24123826  24124826  7728</span>
<span class="co">#&gt; 9   chr1  14293757  14294757  4543</span>
<span class="co">#&gt; 10 chr17  12545783  12546783  4137</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="complement" class="section level3">
<h3>Complement</h3>
<p><code>bed_complement</code> identifies intervals in a genome that are not covered by an input.</p>
<p><img src="img/functions-complement_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human)
<span class="co">#&gt; # A tibble: 723,519 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1     1  1082</span>
<span class="co">#&gt; 2   chr1  2737  3034</span>
<span class="co">#&gt; 3   chr1  4034  6695</span>
<span class="co">#&gt; 4   chr1  7695  8164</span>
<span class="co">#&gt; 5   chr1  9164 18193</span>
<span class="co">#&gt; 6   chr1 19193 19528</span>
<span class="co">#&gt; 7   chr1 20528 25371</span>
<span class="co">#&gt; 8   chr1 26371 28625</span>
<span class="co">#&gt; 9   chr1 29957 30876</span>
<span class="co">#&gt; 10  chr1 31876 34600</span>
<span class="co">#&gt; # ... with 723,509 more rows</span></code></pre></div>
</div>
<div id="merge" class="section level3">
<h3>Merge</h3>
<p><code>bed_merge</code> identifies overlapping intervals and reports new merged ones. <code>is_merged</code> asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.</p>
<p><img src="img/functions-merge_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="fl">1e6</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> n)
<span class="kw"><a href="../reference/is_merged.html">is_merged</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

<span class="co"># add some signal</span>
x &lt;-<span class="st"> </span>x %&gt;%<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">signal =</span> <span class="kw">runif</span>(n))

<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x, <span class="dt">maxs =</span> <span class="kw">max</span>(signal))
<span class="co">#&gt; # A tibble: 723,621 &times; 4</span>
<span class="co">#&gt;    chrom start   end signal</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  1566  2566      .</span>
<span class="co">#&gt; 2   chr1  2733  4715      .</span>
<span class="co">#&gt; 3   chr1  5187  6187      .</span>
<span class="co">#&gt; 4   chr1  7309  8516      .</span>
<span class="co">#&gt; 5   chr1  9356 11005      .</span>
<span class="co">#&gt; 6   chr1 12246 14095      .</span>
<span class="co">#&gt; 7   chr1 17052 18052      .</span>
<span class="co">#&gt; 8   chr1 23445 25381      .</span>
<span class="co">#&gt; 9   chr1 30931 31931      .</span>
<span class="co">#&gt; 10  chr1 34958 36781      .</span>
<span class="co">#&gt; # ... with 723,611 more rows</span></code></pre></div>
</div>
<div id="flank" class="section level3">
<h3>Flank</h3>
<p><code>bed_flank</code> creates new intervals that flank &ndash; but do not contain &ndash; the input intervals.</p>
<p><img src="img/functions-flank_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_flank.html">bed_flank</a></span>(x, genome_human, <span class="dt">both  =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    50   150</span>
<span class="co">#&gt; 2  chr1   125   225</span></code></pre></div>
</div>
<div id="slop" class="section level3">
<h3>Slop</h3>
<p><code>bed_slop</code> pads input intervals based on a specification</p>
<p><img src="img/functions-slop_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_slop.html">bed_slop</a></span>(x, genome_human, <span class="dt">both =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    10   220</span>
<span class="co">#&gt; 2  chr1   125   335</span></code></pre></div>
</div>
<div id="shift" class="section level3">
<h3>Shift</h3>
<p><code>bed_shift</code> adjusts coordinates toward <code>start</code> or <code>end</code> by a defined <code>size</code>. Intervals created out of bounds are removed, or trimmed.</p>
<p><img src="img/functions-shift_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_shift.html">bed_shift</a></span>(x, genome_human, <span class="dt">size =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1   125   150</span>
<span class="co">#&gt; 2  chr1   200   225</span></code></pre></div>
</div>
<div id="interval-spacing" class="section level3">
<h3>Interval spacing</h3>
<p>Interval spacing is easily computed using <code>dplyr</code>. Overlapping intervals must first be merged. Spacing for the first interval of each chromosome is undefined.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x)

x %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">.spacing =</span> start -<span class="st"> </span><span class="kw">lag</span>(end))
<span class="co">#&gt; Source: local data frame [723,957 x 4]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start   end .spacing</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  1116  2116       NA</span>
<span class="co">#&gt; 2   chr1  5555  6555     3439</span>
<span class="co">#&gt; 3   chr1 10991 11991     4436</span>
<span class="co">#&gt; 4   chr1 12005 13371       14</span>
<span class="co">#&gt; 5   chr1 14392 16224     1021</span>
<span class="co">#&gt; 6   chr1 16420 17420      196</span>
<span class="co">#&gt; 7   chr1 20708 21708     3288</span>
<span class="co">#&gt; 8   chr1 24605 25605     2897</span>
<span class="co">#&gt; 9   chr1 25671 26671       66</span>
<span class="co">#&gt; 10  chr1 27553 28807      882</span>
<span class="co">#&gt; # ... with 723,947 more rows</span></code></pre></div>
</div>
</div>
<div id="multiple-set-operations" class="section level2 tabset">
<h2>Multiple set operations</h2>
<p>These methods compare two sets of intervals:</p>
<ul><li><code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>: find overlapping intervals</li>
<li><code><a href="../reference/bed_map.html">bed_map()</a></code>: apply a function to selected columns for overlapping intervals</li>
<li><code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>: Remove intervals based on overlaps between two files</li>
<li><code><a href="../reference/bed_window.html">bed_window()</a></code>: Find overlapping intervals within a window</li>
<li><code><a href="../reference/bed_closest.html">bed_closest()</a></code>: find the closest intervals independent of overlaps</li>
</ul><div id="intersection" class="section level3">
<h3>Intersection</h3>
<p><code><a href="../reference/bed_intersect.html">bed_intersect()</a></code> is implemented using an <a href="https://github.com/ekg/intervaltree/">interval tree</a> in <code>Rcpp</code>. Column names in the result have <code>.x</code> and <code>.y</code> suffixes, and an <code>.overlap</code> column contains the size of the intersection (values of <code>0</code> indicate book-ended, or touching intervals). See the <a href="#benchmarks">benchmarks</a> section for timing. Though <code>bed_intersect</code> is pretty fast already, we intend to further improve upon this by parallization with <a href="https://rcppcore.github.io/RcppParallel/"><code>RcppParallel</code></a>.</p>
<p><img src="img/functions-intersect_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># intersect two sets of 1e6 intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 646,013 &times; 6</span>
<span class="co">#&gt;    chrom start.x end.x start.y end.y .overlap</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt; &lt;int&gt;   &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   13760 14760   13572 14572      812</span>
<span class="co">#&gt; 2   chr1   13760 14760   14175 15175      585</span>
<span class="co">#&gt; 3   chr1   13760 14760   14543 15543      217</span>
<span class="co">#&gt; 4   chr1   15117 16117   14175 15175       58</span>
<span class="co">#&gt; 5   chr1   15117 16117   14543 15543      426</span>
<span class="co">#&gt; 6   chr1   15214 16214   14543 15543      329</span>
<span class="co">#&gt; 7   chr1   33426 34426   32853 33853      427</span>
<span class="co">#&gt; 8   chr1   33813 34813   32853 33853       40</span>
<span class="co">#&gt; 9   chr1   50899 51899   51400 52400      499</span>
<span class="co">#&gt; 10  chr1   53221 54221   52747 53747      526</span>
<span class="co">#&gt; # ... with 646,003 more rows</span>

<span class="co"># A records with no overlaps (i.e., `-v`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y, <span class="dt">invert =</span> <span class="ot">TRUE</span>)
<span class="co">#&gt; # A tibble: 523,930 &times; 3</span>
<span class="co">#&gt;    chrom    start      end</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chrY 59369894 59370894</span>
<span class="co">#&gt; 2   chrY 59343947 59344947</span>
<span class="co">#&gt; 3   chrY 59339555 59340555</span>
<span class="co">#&gt; 4   chrY 59336845 59337845</span>
<span class="co">#&gt; 5   chrY 59330509 59331509</span>
<span class="co">#&gt; 6   chrY 59323701 59324701</span>
<span class="co">#&gt; 7   chrY 59310330 59311330</span>
<span class="co">#&gt; 8   chrY 59308329 59309329</span>
<span class="co">#&gt; 9   chrY 59298269 59299269</span>
<span class="co">#&gt; 10  chrY 59293786 59294786</span>
<span class="co">#&gt; # ... with 523,920 more rows</span></code></pre></div>
<p>One can achieve behaviour similar to BEDtools by combining <code>bed_intersect</code> with <code>dplyr</code> tools.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># `x` records with overlaps (i.e., `-wa`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x)
<span class="co">#&gt; # A tibble: 646,013 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 13760 14760</span>
<span class="co">#&gt; 2   chr1 13760 14760</span>
<span class="co">#&gt; 3   chr1 13760 14760</span>
<span class="co">#&gt; 4   chr1 15117 16117</span>
<span class="co">#&gt; 5   chr1 15117 16117</span>
<span class="co">#&gt; 6   chr1 15214 16214</span>
<span class="co">#&gt; 7   chr1 33426 34426</span>
<span class="co">#&gt; 8   chr1 33813 34813</span>
<span class="co">#&gt; 9   chr1 50899 51899</span>
<span class="co">#&gt; 10  chr1 53221 54221</span>
<span class="co">#&gt; # ... with 646,003 more rows</span>

<span class="co"># `y` records with overlaps (i.e., `-wb`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.y, <span class="dt">end =</span> end.y)
<span class="co">#&gt; # A tibble: 646,013 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 13572 14572</span>
<span class="co">#&gt; 2   chr1 14175 15175</span>
<span class="co">#&gt; 3   chr1 14543 15543</span>
<span class="co">#&gt; 4   chr1 14175 15175</span>
<span class="co">#&gt; 5   chr1 14543 15543</span>
<span class="co">#&gt; 6   chr1 14543 15543</span>
<span class="co">#&gt; 7   chr1 32853 33853</span>
<span class="co">#&gt; 8   chr1 32853 33853</span>
<span class="co">#&gt; 9   chr1 51400 52400</span>
<span class="co">#&gt; 10  chr1 52747 53747</span>
<span class="co">#&gt; # ... with 646,003 more rows</span>

<span class="co"># Unique records in `x` (i.e., `-u`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x) %&gt;%<span class="st"> </span><span class="kw">unique</span>()
<span class="co">#&gt; # A tibble: 475,993 &times; 3</span>
<span class="co">#&gt;    chrom  start    end</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  13760  14760</span>
<span class="co">#&gt; 2   chr1  15117  16117</span>
<span class="co">#&gt; 3   chr1  15214  16214</span>
<span class="co">#&gt; 4   chr1  33426  34426</span>
<span class="co">#&gt; 5   chr1  33813  34813</span>
<span class="co">#&gt; 6   chr1  50899  51899</span>
<span class="co">#&gt; 7   chr1  53221  54221</span>
<span class="co">#&gt; 8   chr1  61934  62934</span>
<span class="co">#&gt; 9   chr1  91859  92859</span>
<span class="co">#&gt; 10  chr1 129346 130346</span>
<span class="co">#&gt; # ... with 475,983 more rows</span>

<span class="co"># Count `y` intervals that intersect each `x` interval</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">group_by</span>(chrom, start.x, end.x) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">count =</span> <span class="kw">n</span>())
<span class="co">#&gt; Source: local data frame [475,993 x 4]</span>
<span class="co">#&gt; Groups: chrom, start.x [?]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start.x  end.x count</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt;  &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   13760  14760     3</span>
<span class="co">#&gt; 2   chr1   15117  16117     2</span>
<span class="co">#&gt; 3   chr1   15214  16214     1</span>
<span class="co">#&gt; 4   chr1   33426  34426     1</span>
<span class="co">#&gt; 5   chr1   33813  34813     1</span>
<span class="co">#&gt; 6   chr1   50899  51899     1</span>
<span class="co">#&gt; 7   chr1   53221  54221     1</span>
<span class="co">#&gt; 8   chr1   61934  62934     1</span>
<span class="co">#&gt; 9   chr1   91859  92859     1</span>
<span class="co">#&gt; 10  chr1  129346 130346     1</span>
<span class="co">#&gt; # ... with 475,983 more rows</span></code></pre></div>
</div>
<div id="map" class="section level3">
<h3>Map</h3>
<p><code>bed_map</code> maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs.</p>
<p><img src="img/functions-map_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bedfile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)
bgfile  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)

x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(bgfile)

<span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">means =</span> <span class="kw">mean</span>(value), <span class="dt">sds =</span> <span class="kw">sd</span>(value))
<span class="co">#&gt; # A tibble: 731 &times; 8</span>
<span class="co">#&gt;    chrom    start      end    means      sds      name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr22 16150259 16193004 17.54545 5.145165      &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 2  chr22 16162065 16172265       NA       NA LINC00516     3      +</span>
<span class="co">#&gt; 3  chr22 16179617 16181004       NA       NA  BC017398     1      -</span>
<span class="co">#&gt; 4  chr22 16239287 16239327       NA       NA  DQ590589     1      +</span>
<span class="co">#&gt; 5  chr22 16240245 16240277       NA       NA  DQ573684     1      -</span>
<span class="co">#&gt; 6  chr22 16240300 16240340       NA       NA  DQ595048     1      -</span>
<span class="co">#&gt; 7  chr22 16241085 16241125       NA       NA  DQ590589     1      +</span>
<span class="co">#&gt; 8  chr22 16242753 16242785       NA       NA  DQ573684     1      -</span>
<span class="co">#&gt; 9  chr22 16243382 16243414       NA       NA  DQ573684     1      -</span>
<span class="co">#&gt; 10 chr22 16243447 16243487       NA       NA  DQ587539     1      -</span>
<span class="co">#&gt; # ... with 721 more rows</span></code></pre></div>
</div>
<div id="subtract" class="section level3">
<h3>Subtract</h3>
<p><code>bed_substract()</code> removes <code>x</code> intervals that intersect with <code>y</code>.</p>
<p><img src="img/functions-subtract_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y)
<span class="co">#&gt; # A tibble: 957,701 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  7040  7996</span>
<span class="co">#&gt; 2   chr1  7392  8392</span>
<span class="co">#&gt; 3   chr1  8164  8762</span>
<span class="co">#&gt; 4   chr1 13154 13412</span>
<span class="co">#&gt; 5   chr1 17755 18755</span>
<span class="co">#&gt; 6   chr1 24535 25297</span>
<span class="co">#&gt; 7   chr1 25436 26436</span>
<span class="co">#&gt; 8   chr1 26102 27102</span>
<span class="co">#&gt; 9   chr1 34984 35984</span>
<span class="co">#&gt; 10  chr1 35118 36118</span>
<span class="co">#&gt; # ... with 957,691 more rows</span></code></pre></div>
</div>
<div id="window" class="section level3">
<h3>Window</h3>
<p><code><a href="../reference/bed_window.html">bed_window()</a></code> identifies <code>y</code> intervals that intersect an expanded window of <code>x</code> intervals.</p>
<p><img src="img/functions-window_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="co"># a few intersections</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span>

<span class="co"># can be expanded by casting a wider net</span>
<span class="kw"><a href="../reference/bed_window.html">bed_window</a></span>(x, y, genome, <span class="dt">both =</span> <span class="fl">1e6</span>)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span></code></pre></div>
</div>
<div id="closest" class="section level3">
<h3>Closest</h3>
<p><code><a href="../reference/bed_closest.html">bed_closest()</a></code> identifies <code>y</code> intervals that are closest to <code>x</code>.</p>
<p><img src="img/functions-closest_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y)
<span class="co">#&gt; # A tibble: 99 &times; 7</span>
<span class="co">#&gt;    chrom   start.x     end.x   start.y     end.y .overlap .distance</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;    &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   3432167   3433167  23931737  23932737        0  20498570</span>
<span class="co">#&gt; 2   chr1 105290625 105291625 148327644 148328644        0  43036019</span>
<span class="co">#&gt; 3   chr1 127112748 127113748 148327644 148328644        0  21213896</span>
<span class="co">#&gt; 4   chr1 154617250 154618250 148327644 148328644        0  -6288606</span>
<span class="co">#&gt; 5   chr1 193731855 193732855 185859168 185860168        0  -7871687</span>
<span class="co">#&gt; 6  chr10    337927    338927  58589860  58590860        0  58250933</span>
<span class="co">#&gt; 7  chr10  24206254  24207254  58589860  58590860        0  34382606</span>
<span class="co">#&gt; 8  chr10  60629142  60630142  58589860  58590860        0  -2038282</span>
<span class="co">#&gt; 9  chr10  72515059  72516059  58589860  58590860        0 -13924199</span>
<span class="co">#&gt; 10 chr10 112137661 112138661 113592013 113593013        0   1453352</span>
<span class="co">#&gt; # ... with 89 more rows</span></code></pre></div>
</div>
</div>
<div id="randomzing-intervals" class="section level2 tabset">
<h2>Randomzing intervals</h2>
<p><code>valr</code> provides methods for creating new random intervals or permutations of existing intervals:</p>
<ul><li><code>bed_random</code> generates random intervals from an input <code>genome</code>.</li>
<li><code>bed_shuffle</code> shuffles coordinates given a set of input intervals.</li>
<li>Random sampling of input intervals is done with <code>dplyr</code>.</li>
</ul><div id="random" class="section level3">
<h3>Random</h3>
<p><code>bed_random</code> generates random intervals from an input genome. The numbers of intervals from each <code>chrom</code> are proporitional to each chrom <code>size</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="fl">1e6</span>, <span class="dt">length =</span> <span class="fl">1e3</span>)
x
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr18  59380095  59381095</span>
<span class="co">#&gt; 2   chrY  54435350  54436350</span>
<span class="co">#&gt; 3   chr3 117104941 117105941</span>
<span class="co">#&gt; 4  chr16  39394526  39395526</span>
<span class="co">#&gt; 5   chrX  50042047  50043047</span>
<span class="co">#&gt; 6  chr10   4248833   4249833</span>
<span class="co">#&gt; 7   chr1 133272167 133273167</span>
<span class="co">#&gt; 8  chr14  33757238  33758238</span>
<span class="co">#&gt; 9  chr17   7143286   7144286</span>
<span class="co">#&gt; 10 chr15  22813856  22814856</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="co"># numbers of sampled intervals are proportional to chrom size</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">n =</span> <span class="kw">n</span>()) %&gt;%<span class="st"> </span><span class="kw">arrange</span>(<span class="kw">desc</span>(n))
<span class="co">#&gt; # A tibble: 25 &times; 2</span>
<span class="co">#&gt;    chrom     n</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 80719</span>
<span class="co">#&gt; 2   chr2 78744</span>
<span class="co">#&gt; 3   chr3 64003</span>
<span class="co">#&gt; 4   chr4 61155</span>
<span class="co">#&gt; 5   chr5 57821</span>
<span class="co">#&gt; 6   chr6 55244</span>
<span class="co">#&gt; 7   chr7 51514</span>
<span class="co">#&gt; 8   chrX 50450</span>
<span class="co">#&gt; 9   chr8 47107</span>
<span class="co">#&gt; 10  chr9 45661</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="sample" class="section level3">
<h3>Sample</h3>
<p>Sampling can be done using <code>dplyr</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="co"># sample by number</span>
<span class="kw">sample_n</span>(x, <span class="fl">1e3</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 1,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr3  29623667  29624667</span>
<span class="co">#&gt; 2   chr8  97753870  97754870</span>
<span class="co">#&gt; 3   chr5  97075421  97076421</span>
<span class="co">#&gt; 4   chrX  98403594  98404594</span>
<span class="co">#&gt; 5  chr10  84565578  84566578</span>
<span class="co">#&gt; 6  chr22  28534134  28535134</span>
<span class="co">#&gt; 7  chr18  56116551  56117551</span>
<span class="co">#&gt; 8  chr16   3416964   3417964</span>
<span class="co">#&gt; 9   chr7  26813243  26814243</span>
<span class="co">#&gt; 10  chr1 135611261 135612261</span>
<span class="co">#&gt; # ... with 990 more rows</span>

<span class="co"># or fraction</span>
<span class="kw">sample_frac</span>(x, <span class="fl">0.1</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 100,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr11  90507593  90508593</span>
<span class="co">#&gt; 2  chr10  70908702  70909702</span>
<span class="co">#&gt; 3   chr3 135969981 135970981</span>
<span class="co">#&gt; 4  chr11  59038625  59039625</span>
<span class="co">#&gt; 5  chr16  60966449  60967449</span>
<span class="co">#&gt; 6  chr12 120374172 120375172</span>
<span class="co">#&gt; 7   chr9  89611071  89612071</span>
<span class="co">#&gt; 8   chr3  56921560  56922560</span>
<span class="co">#&gt; 9  chr11  64363912  64364912</span>
<span class="co">#&gt; 10  chrX  84585454  84586454</span>
<span class="co">#&gt; # ... with 99,990 more rows</span>

<span class="co"># or sample intervals within groups</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">sample_n</span>(<span class="dv">1</span>)
<span class="co">#&gt; Source: local data frame [25 x 3]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 155608945 155609945</span>
<span class="co">#&gt; 2  chr10 114814487 114815487</span>
<span class="co">#&gt; 3  chr11  45974752  45975752</span>
<span class="co">#&gt; 4  chr12  41555996  41556996</span>
<span class="co">#&gt; 5  chr13   2869940   2870940</span>
<span class="co">#&gt; 6  chr14  39367068  39368068</span>
<span class="co">#&gt; 7  chr15  44976112  44977112</span>
<span class="co">#&gt; 8  chr16  41496382  41497382</span>
<span class="co">#&gt; 9  chr17  79870325  79871325</span>
<span class="co">#&gt; 10 chr18  55477233  55478233</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="shuffle" class="section level3">
<h3>Shuffle</h3>
<p><code>bed_shuffle</code> shuffles input intervals. Interval sizes are equal in the input and output.</p>
<pre><code>#&gt; Warning in bind_rows_(x, .id): binding factor and character vector,
#&gt; coercing into character vector</code></pre>
<p><img src="img/functions-shuffle_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human)
y
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;    chrom    start      end</span>
<span class="co">#&gt;   &lt;fctr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr2 27494573 27494593</span>
<span class="co">#&gt; 2   chr2 13630833 13630843</span>

<span class="kw">any</span>(x$start ==<span class="st"> </span>y$start)
<span class="co">#&gt; [1] FALSE</span>

<span class="kw">all</span>(x$end -<span class="st"> </span>x$start ==<span class="st"> </span>y$end -<span class="st"> </span>y$start)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
</div>
<div id="other-topics" class="section level2">
<h2>Other topics</h2>
<div id="grouping-related-intervals" class="section level3">
<h3>Grouping related intervals</h3>
<p>The <code>group_by</code> operation from <code>dplyr</code> serves a similar purpose to BEDtools <code>groupby</code>, and can be combined with <code>dplyr::summarize</code> to calculate summary statistics from groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">overlap.total =</span> <span class="kw">sum</span>(.overlap))
<span class="co">#&gt; # A tibble: 25 &times; 2</span>
<span class="co">#&gt;    chrom overlap.total</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1      25777694</span>
<span class="co">#&gt; 2  chr10      14149118</span>
<span class="co">#&gt; 3  chr11      14067378</span>
<span class="co">#&gt; 4  chr12      13692246</span>
<span class="co">#&gt; 5  chr13      11917205</span>
<span class="co">#&gt; 6  chr14      11068235</span>
<span class="co">#&gt; 7  chr15      10746125</span>
<span class="co">#&gt; 8  chr16       9279278</span>
<span class="co">#&gt; 9  chr17       8644511</span>
<span class="co">#&gt; 10 chr18       8316606</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
</div>
</div>
<div id="interval-statistics" class="section level1 tabset">
<h1>Interval Statistics</h1>
<p><code>valr</code> provides several methods to assess statistical properties of interval sets including:</p>
<ul><li><a href="#bed_fisher"><code><a href="../reference/bed_fisher.html">bed_fisher()</a></code></a>: measure overlap significance of two sets of intervals.</li>
<li><a href="#bed_absdist"><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code></a>: quantify <em>absolute</em> distance between query intervals and closest reference intervals</li>
<li><a href="#bed_reldist"><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code></a>: quantify <em>relative</em> distances between query intervals and closest reference intervals</li>
<li><a href="#bed_jaccard"><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code></a>: quantify extent of overlap between two sets of intervals</li>
</ul><p>Several of these methods were described in the <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">Genometricorr</a> software package.</p>
<div id="fishers-test" class="section level2">
<h2>Fisher&rsquo;s test</h2>
<p>The Fisher&rsquo;s test assesses whether two sets of intervals are drawn from the same background genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_fisher.html">bed_fisher</a></span>(x, y, genome_human)
<span class="co">#&gt;   estimate p.value conf.low conf.high                             method</span>
<span class="co">#&gt; 1 2.339974       0  2.32605  2.354094 Fisher's Exact Test for Count Data</span>
<span class="co">#&gt;   alternative</span>
<span class="co">#&gt; 1   two.sided</span></code></pre></div>
</div>
<div id="absolute-distance" class="section level2">
<h2>Absolute distance</h2>
<p><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code> computes the absolute distance between the midpoint of query intervals and the closest midpoints of a set of reference intervals. Absolute distances are scaled by the inter-reference gap for the chromosome as follows. For <code>Q</code> total query points and <code>R</code> reference points on a chromosome, scale the distance for each query point <code>i</code> to the closest reference point by the inter-reference gap for each chromosome.</p>
<p><span class="math display">\[
d_i(x,y) = min_{\substack{k}}(|q_i - r_k|)\frac{R}{Length\ of\ chromsome}
\]</span></p>
<p>By default both absolute and scaled distances are reported as <code>absdist</code> and <code>scaled_absdist</code> respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_absdist.html">bed_absdist</a></span>(x, y, genome_human)
<span class="co">#&gt; # A tibble: 1,000,000 &times; 5</span>
<span class="co">#&gt;    chrom start   end absdist scaled_absdist</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;   &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1   chr1   888  1888    9540     3.09791830</span>
<span class="co">#&gt; 2   chr1  4570  5570    5858     1.90226472</span>
<span class="co">#&gt; 3   chr1  9262 10262    1166     0.37863446</span>
<span class="co">#&gt; 4   chr1 11827 12827     321     0.10423813</span>
<span class="co">#&gt; 5   chr1 19939 20939     262     0.08507910</span>
<span class="co">#&gt; 6   chr1 21477 22477     323     0.10488759</span>
<span class="co">#&gt; 7   chr1 32269 33269    2257     0.73291421</span>
<span class="co">#&gt; 8   chr1 36693 37693     931     0.30232305</span>
<span class="co">#&gt; 9   chr1 38891 39891    3129     1.01607823</span>
<span class="co">#&gt; 10  chr1 49898 50898      63     0.02045795</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="relative-distance" class="section level2">
<h2>Relative distance</h2>
<p><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> computes the relative distance between a query interval and the two closest reference intervals. The relative distance can range between <code>[0,0.5]</code>. If the overall distribution of distances is biased toward zero, then the query and reference intervals tend to be closer in space (<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">17</a>). By default a summary with the frequency of each relative distance is reported.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_reldist.html">bed_reldist</a></span>(x, y)
<span class="co">#&gt; # A tibble: 51 &times; 4</span>
<span class="co">#&gt;    reldist counts  total       freq</span>
<span class="co">#&gt;      &lt;dbl&gt;  &lt;int&gt;  &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1     0.00  19838 999962 0.01983875</span>
<span class="co">#&gt; 2     0.01  20012 999962 0.02001276</span>
<span class="co">#&gt; 3     0.02  20070 999962 0.02007076</span>
<span class="co">#&gt; 4     0.03  19738 999962 0.01973875</span>
<span class="co">#&gt; 5     0.04  20074 999962 0.02007476</span>
<span class="co">#&gt; 6     0.05  20100 999962 0.02010076</span>
<span class="co">#&gt; 7     0.06  19774 999962 0.01977475</span>
<span class="co">#&gt; 8     0.07  19857 999962 0.01985775</span>
<span class="co">#&gt; 9     0.08  20058 999962 0.02005876</span>
<span class="co">#&gt; 10    0.09  19994 999962 0.01999476</span>
<span class="co">#&gt; # ... with 41 more rows</span></code></pre></div>
</div>
<div id="jaccard-similarity" class="section level2">
<h2>Jaccard similarity</h2>
<p><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code> quantifies the extent of overlap between to sets of intervals. The Jaccard statistic takes values of <code>[0,1]</code> and is measured as:</p>
<p><span class="math display">\[
J(x,y) = \frac{\mid x \bigcap y \mid}
              {\mid x \bigcup y \mid} 
       = 
         \frac{\mid x \bigcap y \mid}
              {\mid x \mid + \mid y \mid - \mid x \bigcap y \mid}
\]</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_jaccard.html">bed_jaccard</a></span>(x, y)
<span class="co">#&gt; # A tibble: 1 &times; 4</span>
<span class="co">#&gt;       len_i      len_u   jaccard      n</span>
<span class="co">#&gt;       &lt;int&gt;      &lt;int&gt;     &lt;dbl&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1 323795233 2000000000 0.1931716 647749</span></code></pre></div>
</div>
</div>
<div id="benchmarks" class="section level1">
<h1>Benchmarks</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># two sets of 1e6 random 1 kb intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw">library</span>(microbenchmark)
<span class="kw">microbenchmark</span>(
  <span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human),
  <span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x),
  <span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human),
  <span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human),
  <span class="dt">times =</span> <span class="dv">1</span>,
  <span class="dt">unit =</span> <span class="st">'s'</span>
)
<span class="co">#&gt; Unit: seconds</span>
<span class="co">#&gt;                             expr       min        lq      mean    median</span>
<span class="co">#&gt;         bed_random(genome_human) 0.1263448 0.1263448 0.1263448 0.1263448</span>
<span class="co">#&gt;                bed_closest(x, y) 3.0438745 3.0438745 3.0438745 3.0438745</span>
<span class="co">#&gt;              bed_intersect(x, y) 1.9584160 1.9584160 1.9584160 1.9584160</span>
<span class="co">#&gt;                     bed_merge(x) 2.6921901 2.6921901 2.6921901 2.6921901</span>
<span class="co">#&gt;               bed_subtract(x, y) 3.7431150 3.7431150 3.7431150 3.7431150</span>
<span class="co">#&gt;  bed_complement(x, genome_human) 3.1129415 3.1129415 3.1129415 3.1129415</span>
<span class="co">#&gt;     bed_shuffle(x, genome_human) 1.2083914 1.2083914 1.2083914 1.2083914</span>
<span class="co">#&gt;         uq       max neval</span>
<span class="co">#&gt;  0.1263448 0.1263448     1</span>
<span class="co">#&gt;  3.0438745 3.0438745     1</span>
<span class="co">#&gt;  1.9584160 1.9584160     1</span>
<span class="co">#&gt;  2.6921901 2.6921901     1</span>
<span class="co">#&gt;  3.7431150 3.7431150     1</span>
<span class="co">#&gt;  3.1129415 3.1129415     1</span>
<span class="co">#&gt;  1.2083914 1.2083914     1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#reading-data">Reading data</a><ul class="nav nav-pills nav-stacked"><li><a href="#bed-files">BED files</a></li>
      <li><a href="#genome-files">Genome files</a></li>
      <li><a href="#bedgraph-files">bedGraph files</a></li>
      <li><a href="#vcf-files">VCF files</a></li>
      </ul></li>
      <li><a href="#interval-manipulations">Interval manipulations</a><ul class="nav nav-pills nav-stacked"><li><a href="#single-set-operations">Single set operations</a></li>
      <li><a href="#multiple-set-operations">Multiple set operations</a></li>
      <li><a href="#randomzing-intervals">Randomzing intervals</a></li>
      <li><a href="#other-topics">Other topics</a></li>
      </ul></li>
      <li><a href="#interval-statistics">Interval Statistics</a><ul class="nav nav-pills nav-stacked"><li><a href="#fishers-test">Fisher&rsquo;s test</a></li>
      <li><a href="#absolute-distance">Absolute distance</a></li>
      <li><a href="#relative-distance">Relative distance</a></li>
      <li><a href="#jaccard-similarity">Jaccard similarity</a></li>
      </ul></li>
      <li><a href="#benchmarks">Benchmarks</a></li>
      </ul></div>
      </div>

</div>


      <footer><p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer></div>

  </body></html>
