<!-- Generated by staticdocs: do not edit by hand -->
<html>
  <head>
  <meta charset="utf-8">
<title><code>valr</code> overview. valr</title>

<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha256-cCueBR6CsyA4/9szpPfrX3s49M9vUU5BgtiJj06wt/s=" crossorigin="anonymous"></script>

<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>


<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">


<!-- staticdocs -->
<link href="../staticdocs.css" rel="stylesheet">

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container">
        <header>
        
<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/valr.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
        
        </header>

        
<div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1><code>valr</code> overview</h1>
                        <h4 class="author"><em>Jay Hesselberth <a href="mailto:jay.hesselberth@gmail.com">jay.hesselberth@gmail.com</a></em></h4>
            
            <h4 class="date"><em>2016-09-28</em></h4>
          </div>

    
    
    <div id="overview" class="section level1">
    <h1>Overview</h1>
    <p><code>valr</code> provides tools to read and manipulate genome intervals and signals, similar to the standalone [<code>BEDtools</code>][1] suite. <code>valr</code> enables BEDtools-like analysis in the R/RStudio environment, and uses modern R tools for a terse, expressive syntax. Compute-intensive algorithms are implemented in [<code>Rcpp</code>][5]/C++, and many methods take advantage of the speed and grouping capability provided by [<code>dplyr</code>][3].</p>
    <div id="installation" class="section level2">
    <h2>Installation</h2>
    <p><code>valr</code> can be installed from github, and will be eventually deposited in CRAN.</p>
    <div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">&#39;jayhesselberth/valr&#39;</span>)</code></pre></div>
    </div>
    <div id="comparison-to-other-tools" class="section level2">
    <h2>Comparison to other tools</h2>
    <p><strong>Why another tool set for interval manipulations?</strong> We already have [BEDtools][1], [bedops][8], [pybedtools][7], [GenomicRanges][9], [bedr][11] and [IRanges][12].</p>
    <p>We were motivated to develop a toolset that:</p>
    <ul>
    <li>Combines analysis and visualization in RStudio.</li>
    <li>Can be used to generate reports with Rmarkdown.</li>
    <li>Is highly extensible. New tools are quickly implemented on the R side.</li>
    <li>Leverages the “modern R” syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code>%&gt;%</code>).</li>
    <li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
    <li>Facilitates interactive visulaizations with [<code>shiny</code>][10].</li>
    </ul>
    <p>We anticipate <code>valr</code> will mainly be used for analysis of pre-processed data in BED, bedGraph and VCF formats. Most users will have processed their aligned reads from BAM format to bedGraph, so we do not foresee supporting BAM directly. We would entertain requests for GTF / GFF support if there is interest, as <code>tidyr</code> makes it easy to convert these to BED12.</p>
    <p>Certain algorithms in <code>valr</code> were implemented in <code>Rcpp</code> (including [intersect, merge, subtract, closest][14]) to enable fluid interactive analysis. See the <a href="#benchmarks">benchmarks</a> section for details.</p>
    </div>
    <div id="why-valr" class="section level2">
    <h2>Why <code>valr</code>?</h2>
    <div id="familiar-tools-all-within-r" class="section level3">
    <h3>Familiar tools, all within R</h3>
    </div>
    <div id="reproducible-reports" class="section level3">
    <h3>Reproducible reports</h3>
    <p><code>valr</code> can be used in RMarkdown documents to generate reproducible work-flows for data processing.</p>
    <p><code>pybedtools</code> can be used within <code>ipython notebooks</code> to accomplish a similar goal, but others have pointed out [issues with this approach][<a href="https://www.r-bloggers.com/why-i-dont-like-jupyter-fka-ipython-notebook/" class="uri">https://www.r-bloggers.com/why-i-dont-like-jupyter-fka-ipython-notebook/</a>].</p>
    </div>
    <div id="non-standard-evaluation" class="section level3">
    <h3>Non-standard evaluation</h3>
    <p>Columns in <code>BEDtools</code> are referred to by position:</p>
    <div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># calculate the mean of column 6 for intervals in `b` that overlap with `a`</span>
    <span class="kw">bedtools</span> map -a a.bed -b b.bed -c 6 -o mean</code></pre></div>
    <p>Ever inspect a multi-column BED file and count columns on your fingers? Several methods in <code>valr</code> use [non-standard evalution][4]. Columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
    <div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate the mean and variance for the `value` column</span>
    <span class="kw">bed_map</span>(x, y, <span class="dt">mean =</span> <span class="kw">mean</span>(value), <span class="dt">var =</span> <span class="kw">var</span>(value))
    
    <span class="co"># report concatenated and max values for merged intervals</span>
    <span class="kw">bed_merge</span>(x, <span class="dt">concat =</span> <span class="kw">concat</span>(value), <span class="dt">max =</span> <span class="kw">max</span>(value))</code></pre></div>
    </div>
    </div>
    <div id="getting-started" class="section level2">
    <h2>Getting started</h2>
    <p>This demonstration combines several <code>valr</code> functions to analyze the distribution of histone marks surrounding transcription start sites.</p>
    <div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
    <span class="kw">library</span>(dplyr)
    <span class="kw">library</span>(ggplot2)
    
    bedfile &lt;-<span class="st"> </span><span class="kw">valr_example</span>(<span class="st">&#39;genes.hg19.chr22.bed.gz&#39;</span>)
    bgfile  &lt;-<span class="st"> </span><span class="kw">valr_example</span>(<span class="st">&#39;hela.h3k4.chip.bg.gz&#39;</span>)
    genomefile &lt;-<span class="st"> </span><span class="kw">valr_example</span>(<span class="st">&#39;hg19.chrom.sizes.gz&#39;</span>)
    
    genes &lt;-<span class="st"> </span><span class="kw">read_bed</span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
    genome &lt;-<span class="st"> </span><span class="kw">read_genome</span>(genomefile)
    
    <span class="co"># generate 1 bp TSS intervals, `+` strand only</span>
    tss &lt;-<span class="st"> </span>genes %&gt;%
    <span class="st">  </span><span class="kw">filter</span>(strand ==<span class="st"> &#39;+&#39;</span>) %&gt;%
    <span class="st">  </span><span class="kw">mutate</span>(<span class="dt">end =</span> start +<span class="st"> </span><span class="dv">1</span>)
    
    region_size &lt;-<span class="st"> </span><span class="dv">1000</span>
    win_size &lt;-<span class="st"> </span><span class="dv">50</span>
    
    <span class="co"># add slop to the TSS, break into windows and add a group</span>
    x &lt;-<span class="st"> </span>tss %&gt;%
    <span class="st">  </span><span class="kw">bed_slop</span>(genome, <span class="dt">both =</span> region_size) %&gt;%
    <span class="st">  </span><span class="kw">bed_makewindows</span>(genome, win_size) %&gt;%
    <span class="st">  </span><span class="kw">group_by</span>(win_id)
    
    y &lt;-<span class="st"> </span><span class="kw">read_bedgraph</span>(bgfile)
    
    <span class="co"># map signals to TSS regions and calculate summary statistics.</span>
    res &lt;-<span class="st"> </span><span class="kw">bed_map</span>(x, y, <span class="dt">win_sum =</span> <span class="kw">sum</span>(value.y)) %&gt;%
    <span class="st">  </span><span class="kw">summarize</span>(<span class="dt">win_mean =</span> <span class="kw">mean</span>(win_sum), <span class="dt">win_sd =</span> <span class="kw">sd</span>(win_sum))
    
    x_labels &lt;-<span class="st"> </span><span class="kw">seq</span>(-region_size, region_size, <span class="dt">by =</span> win_size *<span class="st"> </span><span class="dv">5</span>)
    x_breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">41</span>, <span class="dt">by =</span> <span class="dv">5</span>)
    sd_limits &lt;-<span class="st"> </span><span class="kw">aes</span>(<span class="dt">ymax =</span> win_mean +<span class="st"> </span>win_sd, <span class="dt">ymin =</span> win_mean -<span class="st"> </span>win_sd)
    
    <span class="kw">ggplot</span>(res, <span class="kw">aes</span>(<span class="dt">x =</span> win_id.x, <span class="dt">y =</span> win_mean)) +
    <span class="st">  </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">geom_pointrange</span>(sd_limits) +<span class="st"> </span>
    <span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">labels =</span> x_labels, <span class="dt">breaks =</span> x_breaks) +<span class="st"> </span>
    <span class="st">  </span><span class="kw">xlab</span>(<span class="st">&#39;Position (bp from TSS)&#39;</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="st">&#39;Signal&#39;</span>) +
    <span class="st">  </span><span class="kw">theme_bw</span>()</code></pre></div>
    <div class="figure" style="text-align: center">
    <img src="/Users/jayhesselberth/devel/valr/vignettes/README-demo-tss-1.png" alt="Human H3K4me3 signal near transcription starts sites" width="384" />
    <p class="caption">
    Human H3K4me3 signal near transcription starts sites
    </p>
    </div>
    </div>
    </div>
  </div>

  <div class="col-md-3 hidden-xs toc">

        <div id="nav">
      <h2>Contents</h2>
      <ul>
      <li><a href="#overview">Overview</a><ul>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#comparison-to-other-tools">Comparison to other tools</a></li>
      <li><a href="#why-valr">Why <code>valr</code>?</a><ul>
      <li><a href="#familiar-tools-all-within-r">Familiar tools, all within R</a></li>
      <li><a href="#reproducible-reports">Reproducible reports</a></li>
      <li><a href="#non-standard-evaluation">Non-standard evaluation</a></li>
      </ul></li>
      <li><a href="#getting-started">Getting started</a></li>
      </ul></li>
      </ul>
    </div>
      </div>

</div>


<script>
$('#nav').affix({
  offset: {
    top: $('#nav').offset().top - 80
  }
});
</script>

        <footer>
        <p>Built by <a href="http://hadley.github.io/staticdocs/">staticdocs</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
        </footer>
   </div>

  </body>
</html>
